{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **Catégorisez automatiquement des questions**\n",
    "\n",
    "## partie 2/? : requête API\n",
    "\n",
    "### <br> Un notebook de requête API, qui récupère 50 questions par requête via le wrapper StackAPI ou bien via Stack Exchange API, et qui stocke les principales caractéristiques de ces questions dans un DataFrame.\n",
    "\n",
    "> &emsp; <br><br>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import datetime\n",
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Requete API, via Stack Exchange API\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "# \"réaliser une requête de 50 questions sur une période définie, contenant le tag “python”\n",
    "# et qui ont un score > 50 (votes), récupérer les données principales de la question\n",
    "# (date, titre, tags, score) dans un DataFrame et les afficher\"\n",
    "\n",
    "# Authentification\n",
    "# Pas nécessaire ici, nous avons droit à 10 000 requêtes(lecture)/j sans OAuth 2.0\n",
    "# L' OAuth 2.0 concerne seulement les requetes en écriture\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'items': [{'included_fields': ['.backoff', '.error_id', '.error_message', '.error_name', '.has_more', '.items', '.quota_max', '.quota_remaining', 'access_token.access_token', 'access_token.account_id', 'access_token.expires_on_date', 'access_token.scope', 'account_merge.merge_date', 'account_merge.new_account_id', 'account_merge.old_account_id', 'achievement.account_id', 'achievement.achievement_type', 'achievement.badge_rank', 'achievement.creation_date', 'achievement.is_unread', 'achievement.link', 'achievement.on_site', 'achievement.reputation_change', 'achievement.title', 'answer.answer_id', 'answer.collectives', 'answer.community_owned_date', 'answer.content_license', 'answer.creation_date', 'answer.is_accepted', 'answer.last_activity_date', 'answer.last_edit_date', 'answer.locked_date', 'answer.owner', 'answer.posted_by_collectives', 'answer.question_id', 'answer.recommendations', 'answer.score', 'article.article_id', 'article.article_type', 'article.creation_date', 'article.last_activity_date', 'article.last_edit_date', 'article.link', 'article.owner', 'article.score', 'article.tags', 'article.title', 'article.view_count', 'badge.award_count', 'badge.badge_id', 'badge.badge_type', 'badge.link', 'badge.name', 'badge.rank', 'badge.user', 'badge_count.bronze', 'badge_count.gold', 'badge_count.silver', 'closed_details.by_users', 'closed_details.description', 'closed_details.on_hold', 'closed_details.original_questions', 'closed_details.reason', 'collective.description', 'collective.external_links', 'collective.link', 'collective.name', 'collective.slug', 'collective.tags', 'collective_external_link.link', 'collective_external_link.type', 'collective_membership.collective', 'collective_membership.role', 'collective_recommendation.collective', 'collective_recommendation.creation_date', 'collective_report.collective', 'collective_report.creation_date', 'collective_report.dimensions', 'collective_report.download_link', 'collective_report.end_date', 'collective_report.included_tags', 'collective_report.metrics', 'collective_report.name', 'collective_report.report_id', 'collective_report.report_type', 'collective_report.start_date', 'collective_report.state', 'comment.comment_id', 'comment.content_license', 'comment.creation_date', 'comment.edited', 'comment.owner', 'comment.parent_comment_id', 'comment.post_id', 'comment.post_state', 'comment.reply_to_user', 'comment.score', 'error.description', 'error.error_id', 'error.error_name', 'event.creation_date', 'event.event_id', 'event.event_type', 'exchanged_access_token.access_token', 'exchanged_access_token.account_id', 'exchanged_access_token.exchange_type', 'exchanged_access_token.expires_on_date', 'exchanged_access_token.original_access_token', 'exchanged_access_token.scope', 'filter.filter', 'filter.filter_type', 'filter.included_fields', 'flag_option.count', 'flag_option.description', 'flag_option.dialog_title', 'flag_option.has_flagged', 'flag_option.is_retraction', 'flag_option.option_id', 'flag_option.requires_comment', 'flag_option.requires_question_id', 'flag_option.requires_site', 'flag_option.sub_options', 'flag_option.title', 'inbox_item.answer_id', 'inbox_item.comment_id', 'inbox_item.creation_date', 'inbox_item.is_unread', 'inbox_item.item_type', 'inbox_item.link', 'inbox_item.question_id', 'inbox_item.site', 'inbox_item.title', 'info.answers_per_minute', 'info.api_revision', 'info.badges_per_minute', 'info.new_active_users', 'info.questions_per_minute', 'info.total_accepted', 'info.total_answers', 'info.total_badges', 'info.total_comments', 'info.total_questions', 'info.total_unanswered', 'info.total_users', 'info.total_votes', 'migration_info.on_date', 'migration_info.other_site', 'migration_info.question_id', 'network_activity.account_id', 'network_activity.activity_type', 'network_activity.api_site_parameter', 'network_activity.badge_id', 'network_activity.creation_date', 'network_activity.description', 'network_activity.link', 'network_activity.post_id', 'network_activity.score', 'network_activity.tags', 'network_activity.title', 'network_post.post_id', 'network_post.post_type', 'network_post.score', 'network_post.title', 'network_user.account_id', 'network_user.answer_count', 'network_user.badge_counts', 'network_user.creation_date', 'network_user.last_access_date', 'network_user.question_count', 'network_user.reputation', 'network_user.site_name', 'network_user.site_url', 'network_user.user_id', 'notice.body', 'notice.creation_date', 'notice.owner_user_id', 'notification.body', 'notification.creation_date', 'notification.is_unread', 'notification.notification_type', 'notification.post_id', 'notification.site', 'original_question.accepted_answer_id', 'original_question.answer_count', 'original_question.question_id', 'original_question.title', 'post.collectives', 'post.content_license', 'post.creation_date', 'post.last_activity_date', 'post.last_edit_date', 'post.link', 'post.owner', 'post.post_id', 'post.post_state', 'post.post_type', 'post.posted_by_collectives', 'post.score', 'privilege.description', 'privilege.reputation', 'privilege.short_description', 'question.accepted_answer_id', 'question.answer_count', 'question.bounty_amount', 'question.bounty_closes_date', 'question.closed_date', 'question.closed_reason', 'question.collectives', 'question.community_owned_date', 'question.content_license', 'question.creation_date', 'question.is_answered', 'question.last_activity_date', 'question.last_edit_date', 'question.link', 'question.locked_date', 'question.migrated_from', 'question.migrated_to', 'question.owner', 'question.post_state', 'question.posted_by_collectives', 'question.protected_date', 'question.question_id', 'question.score', 'question.tags', 'question.title', 'question.view_count', 'question_timeline.comment_id', 'question_timeline.content_license', 'question_timeline.creation_date', 'question_timeline.down_vote_count', 'question_timeline.owner', 'question_timeline.post_id', 'question_timeline.question_id', 'question_timeline.revision_guid', 'question_timeline.timeline_type', 'question_timeline.up_vote_count', 'question_timeline.user', 'related_site.api_site_parameter', 'related_site.name', 'related_site.relation', 'related_site.site_url', 'reputation.on_date', 'reputation.post_id', 'reputation.post_type', 'reputation.reputation_change', 'reputation.user_id', 'reputation.vote_type', 'reputation_history.creation_date', 'reputation_history.post_id', 'reputation_history.reputation_change', 'reputation_history.reputation_history_type', 'reputation_history.user_id', 'revision.comment', 'revision.content_license', 'revision.creation_date', 'revision.is_rollback', 'revision.last_tags', 'revision.last_title', 'revision.post_id', 'revision.post_type', 'revision.revision_guid', 'revision.revision_number', 'revision.revision_type', 'revision.set_community_wiki', 'revision.tags', 'revision.title', 'revision.user', 'search_excerpt.answer_count', 'search_excerpt.answer_id', 'search_excerpt.body', 'search_excerpt.creation_date', 'search_excerpt.equivalent_tag_search', 'search_excerpt.excerpt', 'search_excerpt.has_accepted_answer', 'search_excerpt.is_accepted', 'search_excerpt.is_answered', 'search_excerpt.item_type', 'search_excerpt.last_activity_date', 'search_excerpt.question_id', 'search_excerpt.question_score', 'search_excerpt.score', 'search_excerpt.tags', 'search_excerpt.title', 'shallow_user.accept_rate', 'shallow_user.account_id', 'shallow_user.display_name', 'shallow_user.link', 'shallow_user.profile_image', 'shallow_user.reputation', 'shallow_user.user_id', 'shallow_user.user_type', 'site.aliases', 'site.api_site_parameter', 'site.audience', 'site.closed_beta_date', 'site.favicon_url', 'site.high_resolution_icon_url', 'site.icon_url', 'site.launch_date', 'site.logo_url', 'site.markdown_extensions', 'site.name', 'site.open_beta_date', 'site.related_sites', 'site.site_state', 'site.site_type', 'site.site_url', 'site.styling', 'site.twitter_account', 'styling.link_color', 'styling.tag_background_color', 'styling.tag_foreground_color', 'suggested_edit.approval_date', 'suggested_edit.comment', 'suggested_edit.creation_date', 'suggested_edit.post_id', 'suggested_edit.post_type', 'suggested_edit.proposing_user', 'suggested_edit.rejection_date', 'suggested_edit.suggested_edit_id', 'suggested_edit.tags', 'suggested_edit.title', 'tag.collectives', 'tag.count', 'tag.has_synonyms', 'tag.is_moderator_only', 'tag.is_required', 'tag.name', 'tag.user_id', 'tag_preference.tag_name', 'tag_preference.tag_preference_type', 'tag_preference.user_id', 'tag_score.post_count', 'tag_score.score', 'tag_score.user', 'tag_synonym.applied_count', 'tag_synonym.creation_date', 'tag_synonym.from_tag', 'tag_synonym.last_applied_date', 'tag_synonym.to_tag', 'tag_wiki.body_last_edit_date', 'tag_wiki.excerpt', 'tag_wiki.excerpt_last_edit_date', 'tag_wiki.tag_name', 'top_tag.answer_count', 'top_tag.answer_score', 'top_tag.question_count', 'top_tag.question_score', 'top_tag.tag_name', 'top_tag.user_id', 'user.accept_rate', 'user.account_id', 'user.age', 'user.badge_counts', 'user.collectives', 'user.creation_date', 'user.display_name', 'user.is_employee', 'user.last_access_date', 'user.last_modified_date', 'user.link', 'user.location', 'user.profile_image', 'user.reputation', 'user.reputation_change_day', 'user.reputation_change_month', 'user.reputation_change_quarter', 'user.reputation_change_week', 'user.reputation_change_year', 'user.timed_penalty_date', 'user.user_id', 'user.user_type', 'user.website_url', 'user_timeline.badge_id', 'user_timeline.comment_id', 'user_timeline.creation_date', 'user_timeline.detail', 'user_timeline.post_id', 'user_timeline.post_type', 'user_timeline.suggested_edit_id', 'user_timeline.timeline_type', 'user_timeline.title', 'user_timeline.user_id', 'write_permission.can_add', 'write_permission.can_delete', 'write_permission.can_edit', 'write_permission.max_daily_actions', 'write_permission.min_seconds_between_actions', 'write_permission.object_type', 'write_permission.user_id'], 'filter_type': 'safe', 'filter': 'default'}], 'has_more': False, 'quota_max': 300, 'quota_remaining': 188}\n"
     ]
    }
   ],
   "source": [
    "# RGPD\n",
    "# Nous n'avons aucune raison d'extraire des données personnelles ici, donc nous eviterons non seulement\n",
    "# certaines colonnes evidentes, par exemple :\n",
    "# OwnerUserId, OwnerDisplayName, LastEditorUserId, LastEditorDisplayName, etc...\n",
    "# mais aussi la clé primaire (l'id), qui permettrait de les retrouver\n",
    "# (notre cle primaire, si nous en avons besoin, sera simplement l'index de notre dataframe)\n",
    "\n",
    "# Nous allons donc commencer par créer un filtre.\n",
    "# On peut le faire depuis https://api.stackexchange.com/docs/create-filter,\n",
    "# Ou directement depuis ce notebook, en utilisant par exemple le module requests.\n",
    "\n",
    "api_filter_url = 'https://api.stackexchange.com/2.3/filters/create'\n",
    "filter_params = {\n",
    "    'include': 'creation_date,title,tags,score',\n",
    "    'unsafe': 'false',\n",
    "}\n",
    "\n",
    "response = requests.get(api_filter_url, params=filter_params)\n",
    "data = response.json()\n",
    "print(data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'items': [{'included_fields': ['.backoff', '.error_id', '.error_message', '.error_name', '.has_more', '.items', '.quota_max', '.quota_remaining', 'access_token.access_token', 'access_token.account_id', 'access_token.expires_on_date', 'access_token.scope', 'account_merge.merge_date', 'account_merge.new_account_id', 'account_merge.old_account_id', 'achievement.account_id', 'achievement.achievement_type', 'achievement.badge_rank', 'achievement.creation_date', 'achievement.is_unread', 'achievement.link', 'achievement.on_site', 'achievement.reputation_change', 'achievement.title', 'answer.answer_id', 'answer.collectives', 'answer.community_owned_date', 'answer.content_license', 'answer.creation_date', 'answer.is_accepted', 'answer.last_activity_date', 'answer.last_edit_date', 'answer.locked_date', 'answer.owner', 'answer.posted_by_collectives', 'answer.question_id', 'answer.recommendations', 'answer.score', 'article.article_id', 'article.article_type', 'article.creation_date', 'article.last_activity_date', 'article.last_edit_date', 'article.link', 'article.owner', 'article.score', 'article.tags', 'article.title', 'article.view_count', 'badge.award_count', 'badge.badge_id', 'badge.badge_type', 'badge.link', 'badge.name', 'badge.rank', 'badge.user', 'badge_count.bronze', 'badge_count.gold', 'badge_count.silver', 'closed_details.by_users', 'closed_details.description', 'closed_details.on_hold', 'closed_details.original_questions', 'closed_details.reason', 'collective.description', 'collective.external_links', 'collective.link', 'collective.name', 'collective.slug', 'collective.tags', 'collective_external_link.link', 'collective_external_link.type', 'collective_membership.collective', 'collective_membership.role', 'collective_recommendation.collective', 'collective_recommendation.creation_date', 'collective_report.collective', 'collective_report.creation_date', 'collective_report.dimensions', 'collective_report.download_link', 'collective_report.end_date', 'collective_report.included_tags', 'collective_report.metrics', 'collective_report.name', 'collective_report.report_id', 'collective_report.report_type', 'collective_report.start_date', 'collective_report.state', 'comment.comment_id', 'comment.content_license', 'comment.creation_date', 'comment.edited', 'comment.owner', 'comment.parent_comment_id', 'comment.post_id', 'comment.post_state', 'comment.reply_to_user', 'comment.score', 'error.description', 'error.error_id', 'error.error_name', 'event.creation_date', 'event.event_id', 'event.event_type', 'exchanged_access_token.access_token', 'exchanged_access_token.account_id', 'exchanged_access_token.exchange_type', 'exchanged_access_token.expires_on_date', 'exchanged_access_token.original_access_token', 'exchanged_access_token.scope', 'filter.filter', 'filter.filter_type', 'filter.included_fields', 'flag_option.count', 'flag_option.description', 'flag_option.dialog_title', 'flag_option.has_flagged', 'flag_option.is_retraction', 'flag_option.option_id', 'flag_option.requires_comment', 'flag_option.requires_question_id', 'flag_option.requires_site', 'flag_option.sub_options', 'flag_option.title', 'inbox_item.answer_id', 'inbox_item.comment_id', 'inbox_item.creation_date', 'inbox_item.is_unread', 'inbox_item.item_type', 'inbox_item.link', 'inbox_item.question_id', 'inbox_item.site', 'inbox_item.title', 'info.answers_per_minute', 'info.api_revision', 'info.badges_per_minute', 'info.new_active_users', 'info.questions_per_minute', 'info.total_accepted', 'info.total_answers', 'info.total_badges', 'info.total_comments', 'info.total_questions', 'info.total_unanswered', 'info.total_users', 'info.total_votes', 'migration_info.on_date', 'migration_info.other_site', 'migration_info.question_id', 'network_activity.account_id', 'network_activity.activity_type', 'network_activity.api_site_parameter', 'network_activity.badge_id', 'network_activity.creation_date', 'network_activity.description', 'network_activity.link', 'network_activity.post_id', 'network_activity.score', 'network_activity.tags', 'network_activity.title', 'network_post.post_id', 'network_post.post_type', 'network_post.score', 'network_post.title', 'network_user.account_id', 'network_user.answer_count', 'network_user.badge_counts', 'network_user.creation_date', 'network_user.last_access_date', 'network_user.question_count', 'network_user.reputation', 'network_user.site_name', 'network_user.site_url', 'network_user.user_id', 'notice.body', 'notice.creation_date', 'notice.owner_user_id', 'notification.body', 'notification.creation_date', 'notification.is_unread', 'notification.notification_type', 'notification.post_id', 'notification.site', 'original_question.accepted_answer_id', 'original_question.answer_count', 'original_question.question_id', 'original_question.title', 'post.collectives', 'post.content_license', 'post.creation_date', 'post.last_activity_date', 'post.last_edit_date', 'post.link', 'post.owner', 'post.post_id', 'post.post_state', 'post.post_type', 'post.posted_by_collectives', 'post.score', 'privilege.description', 'privilege.reputation', 'privilege.short_description', 'question.accepted_answer_id', 'question.answer_count', 'question.bounty_amount', 'question.bounty_closes_date', 'question.closed_date', 'question.closed_reason', 'question.collectives', 'question.community_owned_date', 'question.content_license', 'question.creation_date', 'question.is_answered', 'question.last_activity_date', 'question.last_edit_date', 'question.link', 'question.locked_date', 'question.migrated_from', 'question.migrated_to', 'question.owner', 'question.post_state', 'question.posted_by_collectives', 'question.protected_date', 'question.question_id', 'question.score', 'question.tags', 'question.title', 'question.view_count', 'question_timeline.comment_id', 'question_timeline.content_license', 'question_timeline.creation_date', 'question_timeline.down_vote_count', 'question_timeline.owner', 'question_timeline.post_id', 'question_timeline.question_id', 'question_timeline.revision_guid', 'question_timeline.timeline_type', 'question_timeline.up_vote_count', 'question_timeline.user', 'related_site.api_site_parameter', 'related_site.name', 'related_site.relation', 'related_site.site_url', 'reputation.on_date', 'reputation.post_id', 'reputation.post_type', 'reputation.reputation_change', 'reputation.user_id', 'reputation.vote_type', 'reputation_history.creation_date', 'reputation_history.post_id', 'reputation_history.reputation_change', 'reputation_history.reputation_history_type', 'reputation_history.user_id', 'revision.comment', 'revision.content_license', 'revision.creation_date', 'revision.is_rollback', 'revision.last_tags', 'revision.last_title', 'revision.post_id', 'revision.post_type', 'revision.revision_guid', 'revision.revision_number', 'revision.revision_type', 'revision.set_community_wiki', 'revision.tags', 'revision.title', 'revision.user', 'search_excerpt.answer_count', 'search_excerpt.answer_id', 'search_excerpt.body', 'search_excerpt.creation_date', 'search_excerpt.equivalent_tag_search', 'search_excerpt.excerpt', 'search_excerpt.has_accepted_answer', 'search_excerpt.is_accepted', 'search_excerpt.is_answered', 'search_excerpt.item_type', 'search_excerpt.last_activity_date', 'search_excerpt.question_id', 'search_excerpt.question_score', 'search_excerpt.score', 'search_excerpt.tags', 'search_excerpt.title', 'shallow_user.accept_rate', 'shallow_user.account_id', 'shallow_user.display_name', 'shallow_user.link', 'shallow_user.profile_image', 'shallow_user.reputation', 'shallow_user.user_id', 'shallow_user.user_type', 'site.aliases', 'site.api_site_parameter', 'site.audience', 'site.closed_beta_date', 'site.favicon_url', 'site.high_resolution_icon_url', 'site.icon_url', 'site.launch_date', 'site.logo_url', 'site.markdown_extensions', 'site.name', 'site.open_beta_date', 'site.related_sites', 'site.site_state', 'site.site_type', 'site.site_url', 'site.styling', 'site.twitter_account', 'styling.link_color', 'styling.tag_background_color', 'styling.tag_foreground_color', 'suggested_edit.approval_date', 'suggested_edit.comment', 'suggested_edit.creation_date', 'suggested_edit.post_id', 'suggested_edit.post_type', 'suggested_edit.proposing_user', 'suggested_edit.rejection_date', 'suggested_edit.suggested_edit_id', 'suggested_edit.tags', 'suggested_edit.title', 'tag.collectives', 'tag.count', 'tag.has_synonyms', 'tag.is_moderator_only', 'tag.is_required', 'tag.name', 'tag.user_id', 'tag_preference.tag_name', 'tag_preference.tag_preference_type', 'tag_preference.user_id', 'tag_score.post_count', 'tag_score.score', 'tag_score.user', 'tag_synonym.applied_count', 'tag_synonym.creation_date', 'tag_synonym.from_tag', 'tag_synonym.last_applied_date', 'tag_synonym.to_tag', 'tag_wiki.body_last_edit_date', 'tag_wiki.excerpt', 'tag_wiki.excerpt_last_edit_date', 'tag_wiki.tag_name', 'top_tag.answer_count', 'top_tag.answer_score', 'top_tag.question_count', 'top_tag.question_score', 'top_tag.tag_name', 'top_tag.user_id', 'user.accept_rate', 'user.account_id', 'user.age', 'user.badge_counts', 'user.collectives', 'user.creation_date', 'user.display_name', 'user.is_employee', 'user.last_access_date', 'user.last_modified_date', 'user.link', 'user.location', 'user.profile_image', 'user.reputation', 'user.reputation_change_day', 'user.reputation_change_month', 'user.reputation_change_quarter', 'user.reputation_change_week', 'user.reputation_change_year', 'user.timed_penalty_date', 'user.user_id', 'user.user_type', 'user.website_url', 'user_timeline.badge_id', 'user_timeline.comment_id', 'user_timeline.creation_date', 'user_timeline.detail', 'user_timeline.post_id', 'user_timeline.post_type', 'user_timeline.suggested_edit_id', 'user_timeline.timeline_type', 'user_timeline.title', 'user_timeline.user_id', 'write_permission.can_add', 'write_permission.can_delete', 'write_permission.can_edit', 'write_permission.max_daily_actions', 'write_permission.min_seconds_between_actions', 'write_permission.object_type', 'write_permission.user_id'], 'filter_type': 'safe', 'filter': 'default'}], 'has_more': False, 'quota_max': 300, 'quota_remaining': 187}\n"
     ]
    }
   ],
   "source": [
    "# ??\n",
    "\n",
    "api_filter_url = 'https://api.stackexchange.com/2.3/filters/create'\n",
    "filter_params = {\n",
    "    'include': 'creation_date,title,tags,score',\n",
    "    'unsafe': 'false',\n",
    "}\n",
    "\n",
    "response = requests.post(api_filter_url, data=filter_params)\n",
    "data = response.json()\n",
    "print(data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Apparemment la stack exchange API inclue des champs par defaut, la doc conseille de filtrer\n",
    "# manuellement la reponse\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Usage of /questions GET\n",
    "# Gets all the questions on the site.\n",
    "\n",
    "# interessant ?\n",
    "# questions/featured     Get all questions on the site with active bounties.\n",
    "# questions/unanswered   Get all questions the site considers unanswered.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "# On peut utiliser directement l'outil en ligne, qui fournit une reponse de 100 questions max,\n",
    "# format JSON compressé.\n",
    "# exemple de la doc :\n",
    "# /2.3/questions?fromdate=1320969600&todate=1321660800&order=desc&min=10&sort=votes&tagged=skyrim&site=gaming\n",
    "\n",
    "# Avant de récupérer nos questions, nous pouvons d'abord nous faire une idée du nombre de questions\n",
    "# qui satisfont nos critères sur une période donnée.\n",
    "\n",
    "# Pour cela il y a plusieurs stratégies possibles,\n",
    "# on peut par exemple utiiser le filtre \"total\" (built-in)\n",
    "# Ou la méthode/endpoint \"search\" (plus flexible)\n",
    "# Comme nous allons au final récupérer des questions, utilisons directement la méthode questions\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'items': [{'tags': ['python', 'error-handling', 'pip'],\n",
       "   'owner': {'account_id': 23480036,\n",
       "    'reputation': 1614,\n",
       "    'user_id': 17532318,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/d37672946ecec10b2814be93daa910ef?s=256&d=identicon&r=PG',\n",
       "    'display_name': 'Apoliticalboy',\n",
       "    'link': 'https://stackoverflow.com/users/17532318/apoliticalboy'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 224796,\n",
       "   'answer_count': 17,\n",
       "   'score': 150,\n",
       "   'last_activity_date': 1700813401,\n",
       "   'creation_date': 1677700339,\n",
       "   'last_edit_date': 1694084649,\n",
       "   'question_id': 75608323,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/75608323/how-do-i-solve-error-externally-managed-environment-every-time-i-use-pip-3',\n",
       "   'title': 'How do I solve &quot;error: externally-managed-environment&quot; every time I use pip 3?'},\n",
       "  {'tags': ['python', 'pandas', 'dataframe', 'attributeerror'],\n",
       "   'owner': {'account_id': 18651684,\n",
       "    'reputation': 1673,\n",
       "    'user_id': 13595011,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://lh3.googleusercontent.com/a-/AOh14Gja1Bf3zfNP7LY4HSFgp6beG2pul2AVcXIsoLC52g=k-s256',\n",
       "    'display_name': 'Maksimjeet Chowdhary',\n",
       "    'link': 'https://stackoverflow.com/users/13595011/maksimjeet-chowdhary'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 261517,\n",
       "   'protected_date': 1695993982,\n",
       "   'answer_count': 3,\n",
       "   'score': 149,\n",
       "   'last_activity_date': 1699567288,\n",
       "   'creation_date': 1680851159,\n",
       "   'last_edit_date': 1691967377,\n",
       "   'question_id': 75956209,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/75956209/error-dataframe-object-has-no-attribute-append',\n",
       "   'title': 'Error &quot;&#39;DataFrame&#39; object has no attribute &#39;append&#39;&quot;'},\n",
       "  {'tags': ['python', 'flask', 'pytest', 'werkzeug'],\n",
       "   'owner': {'account_id': 20745029,\n",
       "    'reputation': 1501,\n",
       "    'user_id': 15233792,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://i.stack.imgur.com/2ol6b.jpg?s=256&g=1',\n",
       "    'display_name': 'stevezkw',\n",
       "    'link': 'https://stackoverflow.com/users/15233792/stevezkw'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 59292,\n",
       "   'protected_date': 1698079880,\n",
       "   'accepted_answer_id': 77214086,\n",
       "   'answer_count': 6,\n",
       "   'score': 100,\n",
       "   'last_activity_date': 1700174126,\n",
       "   'creation_date': 1696215720,\n",
       "   'last_edit_date': 1700174126,\n",
       "   'question_id': 77213053,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/77213053/why-did-flask-start-failing-with-importerror-cannot-import-name-url-quote-fr',\n",
       "   'title': 'Why did Flask start failing with &quot;ImportError: cannot import name &#39;url_quote&#39; from &#39;werkzeug.urls&#39;&quot;?'},\n",
       "  {'tags': ['python', 'openai-api', 'urllib3'],\n",
       "   'owner': {'account_id': 11229285,\n",
       "    'reputation': 1043,\n",
       "    'user_id': 8237910,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://lh5.googleusercontent.com/-OajeBMxDGMo/AAAAAAAAAAI/AAAAAAAAAmQ/8vm8L0FQK1c/photo.jpg?sz=256',\n",
       "    'display_name': 'prismwatermelon',\n",
       "    'link': 'https://stackoverflow.com/users/8237910/prismwatermelon'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 191532,\n",
       "   'accepted_answer_id': 76187415,\n",
       "   'answer_count': 12,\n",
       "   'score': 94,\n",
       "   'last_activity_date': 1699962924,\n",
       "   'creation_date': 1683349902,\n",
       "   'last_edit_date': 1691972425,\n",
       "   'question_id': 76187256,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/76187256/importerror-urllib3-v2-0-only-supports-openssl-1-1-1-currently-the-ssl-modu',\n",
       "   'title': 'ImportError: urllib3 v2.0 only supports OpenSSL 1.1.1+, currently the &#39;ssl&#39; module is compiled with LibreSSL 2.8.3'},\n",
       "  {'tags': ['python', 'prompt', 'openai-api', 'completion', 'chatgpt-api'],\n",
       "   'owner': {'account_id': 3235065,\n",
       "    'reputation': 1732,\n",
       "    'user_id': 3018860,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 89,\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/edec84abcde58e8f5d45a7e5522ff1e1?s=256&d=identicon&r=PG&f=y&so-version=2',\n",
       "    'display_name': 'Unix',\n",
       "    'link': 'https://stackoverflow.com/users/3018860/unix'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 203761,\n",
       "   'closed_date': 1685124780,\n",
       "   'accepted_answer_id': 75898717,\n",
       "   'answer_count': 5,\n",
       "   'score': 94,\n",
       "   'last_activity_date': 1697273724,\n",
       "   'creation_date': 1680263884,\n",
       "   'last_edit_date': 1691966958,\n",
       "   'question_id': 75898276,\n",
       "   'link': 'https://stackoverflow.com/questions/75898276/openai-chatgpt-gpt-3-5-api-error-429-you-exceeded-your-current-quota-please',\n",
       "   'closed_reason': 'Not suitable for this site',\n",
       "   'title': 'OpenAI ChatGPT (GPT-3.5) API error 429: &quot;You exceeded your current quota, please check your plan and billing details&quot;'},\n",
       "  {'tags': ['python', 'ubuntu', 'pip', 'wsl-2'],\n",
       "   'owner': {'account_id': 7752222,\n",
       "    'reputation': 997,\n",
       "    'user_id': 6799513,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://i.stack.imgur.com/Bha40.jpg?s=256&g=1',\n",
       "    'display_name': 'patrick',\n",
       "    'link': 'https://stackoverflow.com/users/6799513/patrick'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 75721,\n",
       "   'accepted_answer_id': 75053968,\n",
       "   'answer_count': 3,\n",
       "   'score': 79,\n",
       "   'last_activity_date': 1687280443,\n",
       "   'creation_date': 1672657056,\n",
       "   'last_edit_date': 1672657567,\n",
       "   'question_id': 74981558,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/74981558/error-updating-python3-pip-attributeerror-module-lib-has-no-attribute-openss',\n",
       "   'title': 'Error Updating Python3 pip AttributeError: module &#39;lib&#39; has no attribute &#39;OpenSSL_add_all_algorithms&#39;'},\n",
       "  {'tags': ['python', 'ffmpeg', 'discord', 'youtube-dl'],\n",
       "   'owner': {'account_id': 27587346,\n",
       "    'reputation': 778,\n",
       "    'user_id': 21055247,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://lh3.googleusercontent.com/a/AEdFTp68sX4uwH8oI00wq1qnsHwmhUC3m9kFQUea4JSE=k-s256',\n",
       "    'display_name': 'nikita goncharov',\n",
       "    'link': 'https://stackoverflow.com/users/21055247/nikita-goncharov'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 88599,\n",
       "   'accepted_answer_id': 75504772,\n",
       "   'answer_count': 9,\n",
       "   'score': 77,\n",
       "   'last_activity_date': 1700757654,\n",
       "   'creation_date': 1676747963,\n",
       "   'last_edit_date': 1700757654,\n",
       "   'question_id': 75495800,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/75495800/error-unable-to-extract-uploader-id-youtube-discord-py',\n",
       "   'title': 'Error: Unable to extract uploader id - Youtube, Discord.py'},\n",
       "  {'tags': ['python', 'pre-commit', 'pre-commit.com', 'isort'],\n",
       "   'owner': {'account_id': 8247924,\n",
       "    'reputation': 5955,\n",
       "    'user_id': 6203472,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/6854a8b5ed6ad561c0ab878c7558fb26?s=256&d=identicon&r=PG&f=y&so-version=2',\n",
       "    'display_name': 'bagerard',\n",
       "    'link': 'https://stackoverflow.com/users/6203472/bagerard'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 13024,\n",
       "   'accepted_answer_id': 75269701,\n",
       "   'answer_count': 5,\n",
       "   'score': 70,\n",
       "   'last_activity_date': 1678725721,\n",
       "   'creation_date': 1674928449,\n",
       "   'last_edit_date': 1675093819,\n",
       "   'question_id': 75269700,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/75269700/pre-commit-fails-to-install-isort-5-11-4-with-error-runtimeerror-the-poetry-co',\n",
       "   'title': 'pre-commit fails to install isort 5.11.4 with error &quot;RuntimeError: The Poetry configuration is invalid&quot;'},\n",
       "  {'tags': ['python', 'python-poetry'],\n",
       "   'owner': {'account_id': 5622044,\n",
       "    'reputation': 7934,\n",
       "    'user_id': 4451521,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 34,\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/4b19b54142d820aad565ca95f920c11c?s=256&d=identicon&r=PG',\n",
       "    'display_name': 'KansaiRobot',\n",
       "    'link': 'https://stackoverflow.com/users/4451521/kansairobot'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 35904,\n",
       "   'answer_count': 4,\n",
       "   'score': 62,\n",
       "   'last_activity_date': 1676545701,\n",
       "   'creation_date': 1675940902,\n",
       "   'question_id': 75397736,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/75397736/poetry-install-on-an-existing-project-error-does-not-contain-any-element',\n",
       "   'title': 'Poetry install on an existing project Error &quot;does not contain any element&quot;'},\n",
       "  {'tags': ['python', 'arrays', 'performance', 'cpython', 'python-internals'],\n",
       "   'owner': {'account_id': 17476274,\n",
       "    'reputation': 24756,\n",
       "    'user_id': 12671057,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/90d2e6942c69df199a95d06e0c8e9a10?s=256&d=identicon&r=PG&f=y&so-version=2',\n",
       "    'display_name': 'Kelly Bundy',\n",
       "    'link': 'https://stackoverflow.com/users/12671057/kelly-bundy'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 6095,\n",
       "   'protected_date': 1676116638,\n",
       "   'accepted_answer_id': 75354915,\n",
       "   'answer_count': 2,\n",
       "   'score': 57,\n",
       "   'last_activity_date': 1676242955,\n",
       "   'creation_date': 1675620257,\n",
       "   'last_edit_date': 1675774367,\n",
       "   'question_id': 75354384,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/75354384/why-is-b-pop0-over-200-times-slower-than-del-b0-for-bytearray',\n",
       "   'title': 'Why is b.pop(0) over 200 times slower than del b[0] for bytearray?'}],\n",
       " 'has_more': False,\n",
       " 'quota_max': 300,\n",
       " 'quota_remaining': 187}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of questions with the tag \"python\" between 2023-01-01 00:00:00 and 2023-11-27 00:00:00 and a minimum score of 50: \n",
      "10\n"
     ]
    }
   ],
   "source": [
    "# All dates in the API are in unix epoch time\n",
    "\n",
    "def get_number_of_questions(tag, start_date, end_date, min_score=50):\n",
    "    api_url = 'https://api.stackexchange.com/2.3/questions'\n",
    "\n",
    "    params = {\n",
    "        'order': 'desc',\n",
    "        'sort': 'votes',\n",
    "        'tagged': tag,\n",
    "        'site': 'stackoverflow',\n",
    "        'fromdate': int(start_date.timestamp()),\n",
    "        'todate': int(end_date.timestamp()),\n",
    "        'min': min_score,\n",
    "        'pagesize': 100  # 30 par défaut, 100 max\n",
    "    }\n",
    "\n",
    "    response = requests.get(api_url, params=params)\n",
    "    data = response.json()\n",
    "    display(data)\n",
    "\n",
    "    if 'items' in data:\n",
    "        return len(data['items'])\n",
    "    else:\n",
    "        return 0\n",
    "\n",
    "# Observons l'activité cette année\n",
    "tag = 'python'\n",
    "start_date = datetime.datetime(2023, 1, 1)\n",
    "end_date = datetime.datetime(2023, 11, 27)\n",
    "\n",
    "result = get_number_of_questions(tag, start_date, end_date)\n",
    "print(f'Number of questions with the tag \"{tag}\" between {start_date} and {end_date} and a minimum score of 50: \\n{result}')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'items': [{'tags': ['python',\n",
       "    'macos',\n",
       "    'terminal',\n",
       "    'atom-editor',\n",
       "    'macos-monterey'],\n",
       "   'owner': {'account_id': 24663681,\n",
       "    'reputation': 3226,\n",
       "    'user_id': 18559642,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/13cdab4eb9d9fb5b1c7dece549f96046?s=256&d=identicon&r=PG',\n",
       "    'display_name': 'Kirill Ignatyev',\n",
       "    'link': 'https://stackoverflow.com/users/18559642/kirill-ignatyev'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 370728,\n",
       "   'protected_date': 1673118234,\n",
       "   'accepted_answer_id': 71621142,\n",
       "   'answer_count': 18,\n",
       "   'score': 197,\n",
       "   'last_activity_date': 1675219189,\n",
       "   'creation_date': 1648058522,\n",
       "   'last_edit_date': 1656277606,\n",
       "   'question_id': 71591971,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/71591971/how-can-i-fix-the-zsh-command-not-found-python-error-macos-monterey-12-3',\n",
       "   'title': 'How can I fix the &quot;zsh: command not found: python&quot; error? (macOS Monterey 12.3, Python&amp;nbsp;3.10, Atom IDE, and atom-python-run 0.9.7)'},\n",
       "  {'tags': ['python', 'tensorflow', 'ray'],\n",
       "   'owner': {'account_id': 22989345,\n",
       "    'reputation': 1981,\n",
       "    'user_id': 17115086,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/16c137f7793988efb97ab410f7aec4f6?s=256&d=identicon&r=PG',\n",
       "    'display_name': 'hasu33',\n",
       "    'link': 'https://stackoverflow.com/users/17115086/hasu33'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 274176,\n",
       "   'protected_date': 1689017298,\n",
       "   'answer_count': 16,\n",
       "   'score': 193,\n",
       "   'last_activity_date': 1698317855,\n",
       "   'creation_date': 1653965255,\n",
       "   'last_edit_date': 1656248470,\n",
       "   'question_id': 72441758,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/72441758/typeerror-descriptors-cannot-not-be-created-directly',\n",
       "   'title': 'TypeError: Descriptors cannot not be created directly'},\n",
       "  {'tags': ['python', 'python-black'],\n",
       "   'owner': {'account_id': 13900281,\n",
       "    'reputation': 11703,\n",
       "    'user_id': 10036039,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/8790cbc891c02fffb597317b43a063fa?s=256&d=identicon&r=PG&f=y&so-version=2',\n",
       "    'display_name': 'zerocewl',\n",
       "    'link': 'https://stackoverflow.com/users/10036039/zerocewl'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 79903,\n",
       "   'accepted_answer_id': 71674345,\n",
       "   'answer_count': 6,\n",
       "   'score': 174,\n",
       "   'last_activity_date': 1679491903,\n",
       "   'creation_date': 1648626375,\n",
       "   'last_edit_date': 1651785651,\n",
       "   'question_id': 71673404,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/71673404/importerror-cannot-import-name-unicodefun-from-click',\n",
       "   'title': 'ImportError: cannot import name &#39;_unicodefun&#39; from &#39;click&#39;'},\n",
       "  {'tags': ['python', 'machine-learning', 'package', 'conda', 'python-poetry'],\n",
       "   'owner': {'account_id': 1449869,\n",
       "    'reputation': 2561,\n",
       "    'user_id': 1367124,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 68,\n",
       "    'profile_image': 'https://i.stack.imgur.com/azGrD.jpg?s=256&g=1',\n",
       "    'display_name': 'Seub',\n",
       "    'link': 'https://stackoverflow.com/users/1367124/seub'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 61264,\n",
       "   'accepted_answer_id': 71110028,\n",
       "   'answer_count': 3,\n",
       "   'score': 142,\n",
       "   'last_activity_date': 1699380700,\n",
       "   'creation_date': 1643123383,\n",
       "   'last_edit_date': 1643126235,\n",
       "   'question_id': 70851048,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/70851048/does-it-make-sense-to-use-conda-poetry',\n",
       "   'title': 'Does it make sense to use Conda + Poetry?'},\n",
       "  {'tags': ['python', 'pandas', 'dataframe', 'data-wrangling', 'data-munging'],\n",
       "   'owner': {'account_id': 29727,\n",
       "    'reputation': 4413,\n",
       "    'user_id': 80544,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 79,\n",
       "    'profile_image': 'https://i.stack.imgur.com/HtN07.jpg?s=256&g=1',\n",
       "    'display_name': 'Glenn',\n",
       "    'link': 'https://stackoverflow.com/users/80544/glenn'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 175243,\n",
       "   'accepted_answer_id': 70837545,\n",
       "   'answer_count': 8,\n",
       "   'score': 142,\n",
       "   'last_activity_date': 1690552412,\n",
       "   'creation_date': 1643042804,\n",
       "   'last_edit_date': 1645482240,\n",
       "   'question_id': 70837397,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/70837397/good-alternative-to-pandas-append-method-now-that-it-is-being-deprecated',\n",
       "   'title': 'Good alternative to Pandas .append() method, now that it is being deprecated?'},\n",
       "  {'tags': ['python', 'google-analytics-api'],\n",
       "   'owner': {'account_id': 6468208,\n",
       "    'reputation': 6455,\n",
       "    'user_id': 5009293,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 76,\n",
       "    'profile_image': 'https://graph.facebook.com/995762863775266/picture?type=large',\n",
       "    'display_name': 'Doug Coats',\n",
       "    'link': 'https://stackoverflow.com/users/5009293/doug-coats'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 205518,\n",
       "   'accepted_answer_id': 73932621,\n",
       "   'answer_count': 18,\n",
       "   'score': 128,\n",
       "   'last_activity_date': 1693903144,\n",
       "   'creation_date': 1663949925,\n",
       "   'last_edit_date': 1663952437,\n",
       "   'question_id': 73830524,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/73830524/attributeerror-module-lib-has-no-attribute-x509-v-flag-cb-issuer-check',\n",
       "   'title': 'AttributeError: module &#39;lib&#39; has no attribute &#39;X509_V_FLAG_CB_ISSUER_CHECK&#39;'},\n",
       "  {'tags': ['python',\n",
       "    'docker',\n",
       "    'docker-compose',\n",
       "    'dependencies',\n",
       "    'importerror'],\n",
       "   'owner': {'account_id': 3369545,\n",
       "    'reputation': 3171,\n",
       "    'user_id': 2829150,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 51,\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/1e9c788890b3d2a8e468ed2773864b2c?s=256&d=identicon&r=PG&f=y&so-version=2',\n",
       "    'display_name': 'Arie',\n",
       "    'link': 'https://stackoverflow.com/users/2829150/arie'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 152032,\n",
       "   'protected_date': 1691442712,\n",
       "   'answer_count': 15,\n",
       "   'score': 117,\n",
       "   'last_activity_date': 1699634829,\n",
       "   'creation_date': 1652208276,\n",
       "   'last_edit_date': 1667754370,\n",
       "   'question_id': 72191560,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/72191560/importerror-cannot-import-name-soft-unicode-from-markupsafe',\n",
       "   'title': 'ImportError: cannot import name &#39;soft_unicode&#39; from &#39;markupsafe&#39;'},\n",
       "  {'tags': ['python',\n",
       "    'selenium',\n",
       "    'google-chrome',\n",
       "    'selenium-webdriver',\n",
       "    'selenium-chromedriver'],\n",
       "   'owner': {'account_id': 19488994,\n",
       "    'reputation': 1083,\n",
       "    'user_id': 14257612,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/c49440f3d6f3831211bcb56d9614a61e?s=256&d=identicon&r=PG&f=y&so-version=2',\n",
       "    'display_name': 'Alesss',\n",
       "    'link': 'https://stackoverflow.com/users/14257612/alesss'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 251439,\n",
       "   'accepted_answer_id': 72773269,\n",
       "   'answer_count': 4,\n",
       "   'score': 108,\n",
       "   'last_activity_date': 1700498304,\n",
       "   'creation_date': 1656337711,\n",
       "   'last_edit_date': 1661689426,\n",
       "   'question_id': 72773206,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/72773206/selenium-python-attributeerror-webdriver-object-has-no-attribute-find-el',\n",
       "   'title': 'Selenium - Python - AttributeError: &#39;WebDriver&#39; object has no attribute &#39;find_element_by_name&#39;'},\n",
       "  {'tags': ['python', 'linux', 'openssl', 'python-poetry'],\n",
       "   'owner': {'account_id': 7489950,\n",
       "    'reputation': 4646,\n",
       "    'user_id': 5692012,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 50,\n",
       "    'profile_image': 'https://i.stack.imgur.com/a3FwR.gif?s=256&g=1',\n",
       "    'display_name': 'RobinFrcd',\n",
       "    'link': 'https://stackoverflow.com/users/5692012/robinfrcd'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 193080,\n",
       "   'accepted_answer_id': 72633324,\n",
       "   'answer_count': 14,\n",
       "   'score': 107,\n",
       "   'last_activity_date': 1699036115,\n",
       "   'creation_date': 1651782719,\n",
       "   'last_edit_date': 1675089528,\n",
       "   'question_id': 72133316,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/72133316/libssl-so-1-1-cannot-open-shared-object-file-no-such-file-or-directory',\n",
       "   'title': 'libssl.so.1.1: cannot open shared object file: No such file or directory'},\n",
       "  {'tags': ['python', 'windows', 'visual-studio-2022'],\n",
       "   'owner': {'account_id': 2813476,\n",
       "    'reputation': 1813,\n",
       "    'user_id': 2419205,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 52,\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/5a4b5ff5c0d06ba079e8d974dd8309b4?s=256&d=identicon&r=PG',\n",
       "    'display_name': 'Starfleet Security',\n",
       "    'link': 'https://stackoverflow.com/users/2419205/starfleet-security'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 7130,\n",
       "   'answer_count': 9,\n",
       "   'score': 105,\n",
       "   'last_activity_date': 1668277374,\n",
       "   'creation_date': 1645204638,\n",
       "   'last_edit_date': 1648802107,\n",
       "   'question_id': 71177450,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/71177450/unable-to-restore-visual-studio-2022-window-ide-itself-after-minimizing',\n",
       "   'title': 'Unable to restore Visual Studio 2022 window (IDE itself) after minimizing'},\n",
       "  {'tags': ['python',\n",
       "    'string',\n",
       "    'reference',\n",
       "    'immutability',\n",
       "    'string-concatenation'],\n",
       "   'owner': {'account_id': 6666507,\n",
       "    'reputation': 1245,\n",
       "    'user_id': 5144013,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 44,\n",
       "    'profile_image': 'https://lh3.googleusercontent.com/-gaua46VyxKQ/AAAAAAAAAAI/AAAAAAAAHxg/M-29RPq_C6E/photo.jpg?sz=256',\n",
       "    'display_name': 'Master_Roshy',\n",
       "    'link': 'https://stackoverflow.com/users/5144013/master-roshy'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 6831,\n",
       "   'protected_date': 1657262192,\n",
       "   'answer_count': 4,\n",
       "   'score': 102,\n",
       "   'last_activity_date': 1657261330,\n",
       "   'creation_date': 1654830500,\n",
       "   'last_edit_date': 1654982358,\n",
       "   'question_id': 72568719,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/72568719/if-python-strings-are-immutable-why-does-it-keep-the-same-id-if-i-use-to-app',\n",
       "   'title': 'If Python strings are immutable, why does it keep the same id if I use += to append to it?'},\n",
       "  {'tags': ['python', 'cpython', 'python-internals'],\n",
       "   'owner': {'account_id': 17476274,\n",
       "    'reputation': 24756,\n",
       "    'user_id': 12671057,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/90d2e6942c69df199a95d06e0c8e9a10?s=256&d=identicon&r=PG&f=y&so-version=2',\n",
       "    'display_name': 'Kelly Bundy',\n",
       "    'link': 'https://stackoverflow.com/users/12671057/kelly-bundy'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 8221,\n",
       "   'closed_date': 1643494282,\n",
       "   'answer_count': 3,\n",
       "   'score': 101,\n",
       "   'last_activity_date': 1644084297,\n",
       "   'creation_date': 1643300597,\n",
       "   'last_edit_date': 1644083484,\n",
       "   'question_id': 70882092,\n",
       "   'link': 'https://stackoverflow.com/questions/70882092/can-we-make-1-2-true',\n",
       "   'closed_reason': 'Duplicate',\n",
       "   'title': 'Can we make 1 == 2 true?'},\n",
       "  {'tags': ['python', 'c++', 'python-3.x', 'rounding', 'integer-division'],\n",
       "   'owner': {'account_id': 7923772,\n",
       "    'reputation': 7078,\n",
       "    'user_id': 5983841,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 62,\n",
       "    'profile_image': 'https://i.stack.imgur.com/GHJxG.jpg?s=256&g=1',\n",
       "    'display_name': 'Rick',\n",
       "    'link': 'https://stackoverflow.com/users/5983841/rick'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 8310,\n",
       "   'accepted_answer_id': 70735005,\n",
       "   'answer_count': 8,\n",
       "   'score': 88,\n",
       "   'last_activity_date': 1660790087,\n",
       "   'creation_date': 1642342052,\n",
       "   'last_edit_date': 1643506145,\n",
       "   'question_id': 70730831,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/70730831/whats-the-mathematical-reason-behind-python-choosing-to-round-integer-division',\n",
       "   'title': 'What&#39;s the mathematical reason behind Python choosing to round integer division toward negative infinity?'},\n",
       "  {'tags': ['python', 'pip'],\n",
       "   'owner': {'account_id': 15449541,\n",
       "    'reputation': 973,\n",
       "    'user_id': 17660775,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://graph.facebook.com/2328891030722501/picture?type=large',\n",
       "    'display_name': 'Abdeldjalil Sekkal',\n",
       "    'link': 'https://stackoverflow.com/users/17660775/abdeldjalil-sekkal'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 154116,\n",
       "   'protected_date': 1676571122,\n",
       "   'answer_count': 10,\n",
       "   'score': 82,\n",
       "   'last_activity_date': 1698355404,\n",
       "   'creation_date': 1653937478,\n",
       "   'last_edit_date': 1654222298,\n",
       "   'question_id': 72439001,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/72439001/there-was-an-error-checking-the-latest-version-of-pip',\n",
       "   'title': 'There was an error checking the latest version of pip'},\n",
       "  {'tags': ['python', 'performance', 'comparison'],\n",
       "   'owner': {'account_id': 393493,\n",
       "    'reputation': 4941,\n",
       "    'user_id': 755371,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 44,\n",
       "    'profile_image': 'https://i.stack.imgur.com/Cmp2K.png?s=256&g=1',\n",
       "    'display_name': 'Eric',\n",
       "    'link': 'https://stackoverflow.com/users/755371/eric'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 7881,\n",
       "   'closed_date': 1648640765,\n",
       "   'accepted_answer_id': 71644988,\n",
       "   'answer_count': 2,\n",
       "   'score': 77,\n",
       "   'last_activity_date': 1648641525,\n",
       "   'creation_date': 1648455910,\n",
       "   'last_edit_date': 1648563343,\n",
       "   'question_id': 71644405,\n",
       "   'link': 'https://stackoverflow.com/questions/71644405/why-is-it-faster-to-compare-strings-that-match-than-strings-that-do-not',\n",
       "   'closed_reason': 'Duplicate',\n",
       "   'title': 'Why is it faster to compare strings that match than strings that do not?'},\n",
       "  {'tags': ['python', 'python-3.x', 'pip'],\n",
       "   'owner': {'account_id': 23650540,\n",
       "    'reputation': 805,\n",
       "    'user_id': 17679479,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://lh3.googleusercontent.com/a-/AOh14Gir3m-DTE1Yx2GwtNYJ0w02mRyRj-8yB5ZPmIx5=k-s256',\n",
       "    'display_name': 'BRICK MANE',\n",
       "    'link': 'https://stackoverflow.com/users/17679479/brick-mane'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 142518,\n",
       "   'answer_count': 15,\n",
       "   'score': 74,\n",
       "   'last_activity_date': 1693433857,\n",
       "   'creation_date': 1643729700,\n",
       "   'last_edit_date': 1658994516,\n",
       "   'question_id': 70943244,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/70943244/attributeerror-module-collections-has-no-attribute-mutablemapping',\n",
       "   'title': 'AttributeError: module &#39;collections&#39; has no attribute &#39;MutableMapping&#39;'},\n",
       "  {'tags': ['python', 'django', 'digital-ocean'],\n",
       "   'owner': {'account_id': 8967886,\n",
       "    'reputation': 3682,\n",
       "    'user_id': 6687699,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 58,\n",
       "    'profile_image': 'https://i.stack.imgur.com/4Snxy.jpg?s=256&g=1',\n",
       "    'display_name': 'Lutaaya Huzaifah Idris',\n",
       "    'link': 'https://stackoverflow.com/users/6687699/lutaaya-huzaifah-idris'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 40237,\n",
       "   'accepted_answer_id': 73932581,\n",
       "   'answer_count': 5,\n",
       "   'score': 72,\n",
       "   'last_activity_date': 1675852882,\n",
       "   'creation_date': 1664747002,\n",
       "   'question_id': 73929564,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/73929564/entrypoints-object-has-no-attribute-get-digital-ocean',\n",
       "   'title': '&#39;EntryPoints&#39; object has no attribute &#39;get&#39; - Digital ocean'},\n",
       "  {'tags': ['python', 'enums', 'f-string'],\n",
       "   'owner': {'account_id': 10991861,\n",
       "    'reputation': 6458,\n",
       "    'user_id': 8075540,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://i.stack.imgur.com/0eZhT.jpg?s=256&g=1',\n",
       "    'display_name': 'Daniel Walker',\n",
       "    'link': 'https://stackoverflow.com/users/8075540/daniel-walker'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 8419,\n",
       "   'accepted_answer_id': 72449614,\n",
       "   'answer_count': 2,\n",
       "   'score': 71,\n",
       "   'last_activity_date': 1693593159,\n",
       "   'creation_date': 1654006677,\n",
       "   'last_edit_date': 1660255156,\n",
       "   'question_id': 72449482,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/72449482/f-string-representation-different-than-str',\n",
       "   'title': 'f-string representation different than str()'},\n",
       "  {'tags': ['python', 'docker', 'flask', 'docker-compose'],\n",
       "   'owner': {'account_id': 22335364,\n",
       "    'reputation': 884,\n",
       "    'user_id': 16556045,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://lh3.googleusercontent.com/a-/AOh14GitmoVYWKb7Xize5hPMshSWNoFyMKSumgrMWNBXBA=k-s256',\n",
       "    'display_name': 'KronosHedronos2077',\n",
       "    'link': 'https://stackoverflow.com/users/16556045/kronoshedronos2077'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 73564,\n",
       "   'accepted_answer_id': 71219718,\n",
       "   'answer_count': 11,\n",
       "   'score': 69,\n",
       "   'last_activity_date': 1691078711,\n",
       "   'creation_date': 1645310480,\n",
       "   'last_edit_date': 1663178530,\n",
       "   'question_id': 71189819,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/71189819/importerror-cannot-import-name-json-from-itsdangerous',\n",
       "   'title': 'ImportError: cannot import name &#39;json&#39; from itsdangerous'},\n",
       "  {'tags': ['python', 'ssl', 'openssl', 'runtime-error'],\n",
       "   'owner': {'account_id': 17018517,\n",
       "    'reputation': 781,\n",
       "    'user_id': 12311729,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://graph.facebook.com/2152598004840080/picture?type=large',\n",
       "    'display_name': 'Gianmarco Broilo',\n",
       "    'link': 'https://stackoverflow.com/users/12311729/gianmarco-broilo'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 144014,\n",
       "   'accepted_answer_id': 72374542,\n",
       "   'answer_count': 12,\n",
       "   'score': 67,\n",
       "   'last_activity_date': 1700615763,\n",
       "   'creation_date': 1648128152,\n",
       "   'last_edit_date': 1700615763,\n",
       "   'question_id': 71603314,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/71603314/ssl-error-unsafe-legacy-renegotiation-disabled',\n",
       "   'title': 'SSL error unsafe legacy renegotiation disabled'},\n",
       "  {'tags': ['python', 'python-3.x', 'pycharm'],\n",
       "   'owner': {'account_id': 6823618,\n",
       "    'reputation': 1052,\n",
       "    'user_id': 5250196,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://i.stack.imgur.com/1QajQ.jpg?s=256&g=1',\n",
       "    'display_name': 'P2000',\n",
       "    'link': 'https://stackoverflow.com/users/5250196/p2000'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 5822,\n",
       "   'accepted_answer_id': 72456461,\n",
       "   'answer_count': 1,\n",
       "   'score': 63,\n",
       "   'last_activity_date': 1666193246,\n",
       "   'creation_date': 1654055104,\n",
       "   'last_edit_date': 1654143878,\n",
       "   'question_id': 72456234,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/72456234/why-does-print-i-e-three-dots-in-a-row-print-blank',\n",
       "   'title': 'Why does print(&quot;...&quot;), i.e. three dots in a row, print blank?'},\n",
       "  {'tags': ['python', 'tensorflow', 'conda', 'apple-m1'],\n",
       "   'owner': {'account_id': 114708,\n",
       "    'reputation': 10637,\n",
       "    'user_id': 301513,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 47,\n",
       "    'profile_image': 'https://i.stack.imgur.com/oWFRY.png?s=256&g=1',\n",
       "    'display_name': 'Qiulang',\n",
       "    'link': 'https://stackoverflow.com/users/301513/qiulang'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 63417,\n",
       "   'protected_date': 1694544271,\n",
       "   'answer_count': 14,\n",
       "   'score': 61,\n",
       "   'last_activity_date': 1695005358,\n",
       "   'creation_date': 1657708328,\n",
       "   'last_edit_date': 1695005358,\n",
       "   'question_id': 72964800,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/72964800/what-is-the-proper-way-to-install-tensorflow-on-apple-m1-in-2022',\n",
       "   'title': 'What is the proper way to install TensorFlow on Apple M1 in 2022'},\n",
       "  {'tags': ['python', 'numpy'],\n",
       "   'owner': {'account_id': 12810242,\n",
       "    'reputation': 804,\n",
       "    'user_id': 9481479,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://graph.facebook.com/1769985833305230/picture?type=large',\n",
       "    'display_name': 'Nawin K Sharma',\n",
       "    'link': 'https://stackoverflow.com/users/9481479/nawin-k-sharma'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 87582,\n",
       "   'accepted_answer_id': 74861894,\n",
       "   'answer_count': 7,\n",
       "   'score': 60,\n",
       "   'last_activity_date': 1696327318,\n",
       "   'creation_date': 1671396136,\n",
       "   'last_edit_date': 1677183023,\n",
       "   'question_id': 74844262,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/74844262/how-can-i-solve-error-module-numpy-has-no-attribute-float-in-python',\n",
       "   'title': 'How can I solve error &quot;module &#39;numpy&#39; has no attribute &#39;float&#39;&quot; in Python?'},\n",
       "  {'tags': ['python', 'ipython'],\n",
       "   'owner': {'account_id': 3582471,\n",
       "    'reputation': 2580,\n",
       "    'user_id': 2990344,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 69,\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/a3f9a2172c31b0a956c093d24230facd?s=256&d=identicon&r=PG&f=y&so-version=2',\n",
       "    'display_name': 'xdavidliu',\n",
       "    'link': 'https://stackoverflow.com/users/2990344/xdavidliu'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 8269,\n",
       "   'accepted_answer_id': 71459528,\n",
       "   'answer_count': 3,\n",
       "   'score': 59,\n",
       "   'last_activity_date': 1684946724,\n",
       "   'creation_date': 1644771035,\n",
       "   'question_id': 71102876,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/71102876/in-ipython-how-do-i-accept-and-use-an-autocomplete-suggestion',\n",
       "   'title': 'in ipython how do I accept and use an autocomplete suggestion?'},\n",
       "  {'tags': ['python', 'opencv'],\n",
       "   'owner': {'account_id': 21236531,\n",
       "    'reputation': 561,\n",
       "    'user_id': 15620641,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://lh5.googleusercontent.com/-chjRp4lpZnM/AAAAAAAAAAI/AAAAAAAAAAA/AMZuuckVdG7sDDNo-Buu9gVMDrWohJSEFA/s96-c/photo.jpg?sz=256',\n",
       "    'display_name': '&#214;mer Faruk Aydın',\n",
       "    'link': 'https://stackoverflow.com/users/15620641/%c3%96mer-faruk-ayd%c4%b1n'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 110365,\n",
       "   'answer_count': 8,\n",
       "   'score': 56,\n",
       "   'last_activity_date': 1676875906,\n",
       "   'creation_date': 1655839048,\n",
       "   'question_id': 72706073,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/72706073/attributeerror-partially-initialized-module-cv2-has-no-attribute-gapi-wip-gs',\n",
       "   'title': 'AttributeError: partially initialized module &#39;cv2&#39; has no attribute &#39;gapi_wip_gst_GStreamerPipeline&#39; (most likely due to a circular import)'},\n",
       "  {'tags': ['python', 'c++', 'numpy', 'complex-numbers', 'infinity'],\n",
       "   'owner': {'account_id': 1705643,\n",
       "    'reputation': 928,\n",
       "    'user_id': 1564730,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 75,\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/81f8bd69726693149da7f75f0655e5b4?s=256&d=identicon&r=PG',\n",
       "    'display_name': 'Firman',\n",
       "    'link': 'https://stackoverflow.com/users/1564730/firman'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 3925,\n",
       "   'accepted_answer_id': 74799453,\n",
       "   'answer_count': 3,\n",
       "   'score': 55,\n",
       "   'last_activity_date': 1671443913,\n",
       "   'creation_date': 1671022480,\n",
       "   'last_edit_date': 1671437130,\n",
       "   'question_id': 74798626,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/74798626/why-is-loginf-inf-j-equal-to-inf-0-785398-j-in-c-python-numpy',\n",
       "   'title': 'Why is log(inf + inf j) equal to (inf + 0.785398 j), In C++/Python/NumPy?'},\n",
       "  {'tags': ['python', 'django', 'heroku'],\n",
       "   'owner': {'account_id': 16660311,\n",
       "    'reputation': 640,\n",
       "    'user_id': 12040445,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://i.stack.imgur.com/ydOnG.jpg?s=256&g=1',\n",
       "    'display_name': 'SarveshJoshi',\n",
       "    'link': 'https://stackoverflow.com/users/12040445/sarveshjoshi'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 82985,\n",
       "   'accepted_answer_id': 72796492,\n",
       "   'answer_count': 11,\n",
       "   'score': 54,\n",
       "   'last_activity_date': 1697317485,\n",
       "   'creation_date': 1648843469,\n",
       "   'question_id': 71712258,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/71712258/error-could-not-build-wheels-for-backports-zoneinfo-which-is-required-to-insta',\n",
       "   'title': 'ERROR: Could not build wheels for backports.zoneinfo, which is required to install pyproject.toml-based projects'},\n",
       "  {'tags': ['python', 'flask', 'jinja2'],\n",
       "   'owner': {'account_id': 15865542,\n",
       "    'reputation': 1201,\n",
       "    'user_id': 11447688,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/16e8410a6faec497b5f4eae530cda0a7?s=256&d=identicon&r=PG&f=y&so-version=2',\n",
       "    'display_name': 'young_minds1',\n",
       "    'link': 'https://stackoverflow.com/users/11447688/young-minds1'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 122447,\n",
       "   'answer_count': 12,\n",
       "   'score': 54,\n",
       "   'last_activity_date': 1691384656,\n",
       "   'creation_date': 1648460099,\n",
       "   'last_edit_date': 1651468617,\n",
       "   'question_id': 71645272,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/71645272/importerror-cannot-import-name-markup-from-jinja2',\n",
       "   'title': 'ImportError: cannot import name &#39;Markup&#39; from &#39;jinja2&#39;'},\n",
       "  {'tags': ['python',\n",
       "    'floating-point',\n",
       "    'complex-numbers',\n",
       "    'approximate',\n",
       "    'structural-pattern-matching'],\n",
       "   'owner': {'account_id': 186862,\n",
       "    'reputation': 218388,\n",
       "    'user_id': 424499,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/bc2071afd499daef001e75e14d7f9cce?s=256&d=identicon&r=PG',\n",
       "    'display_name': 'Raymond Hettinger',\n",
       "    'link': 'https://stackoverflow.com/users/424499/raymond-hettinger'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 3354,\n",
       "   'accepted_answer_id': 72596437,\n",
       "   'answer_count': 2,\n",
       "   'score': 54,\n",
       "   'last_activity_date': 1656679394,\n",
       "   'creation_date': 1655074449,\n",
       "   'last_edit_date': 1655075960,\n",
       "   'question_id': 72596436,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/72596436/how-to-perform-approximate-structural-pattern-matching-for-floats-and-complex',\n",
       "   'title': 'How to perform approximate structural pattern matching for floats and complex'},\n",
       "  {'tags': ['python', 'pip', 'setuptools', 'pyproject.toml'],\n",
       "   'owner': {'account_id': 2017134,\n",
       "    'reputation': 24117,\n",
       "    'user_id': 1804173,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 77,\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/53f721a451a1c68caf8316cefcde3083?s=256&d=identicon&r=PG',\n",
       "    'display_name': 'bluenote10',\n",
       "    'link': 'https://stackoverflow.com/users/1804173/bluenote10'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 22031,\n",
       "   'accepted_answer_id': 73600610,\n",
       "   'answer_count': 2,\n",
       "   'score': 54,\n",
       "   'last_activity_date': 1694478158,\n",
       "   'creation_date': 1662300788,\n",
       "   'last_edit_date': 1667985506,\n",
       "   'question_id': 73600082,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/73600082/how-to-reference-a-requirements-txt-in-the-pyproject-toml-of-a-setuptools-projec',\n",
       "   'title': 'How to reference a requirements.txt in the pyproject.toml of a setuptools project?'},\n",
       "  {'tags': ['python', 'python-typing'],\n",
       "   'owner': {'account_id': 1881254,\n",
       "    'reputation': 1102,\n",
       "    'user_id': 1701415,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/cb5160e829bd4a3a0863e8bd6cff9aea?s=256&d=identicon&r=PG',\n",
       "    'display_name': 'Jesse',\n",
       "    'link': 'https://stackoverflow.com/users/1701415/jesse'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 27181,\n",
       "   'accepted_answer_id': 76399911,\n",
       "   'answer_count': 4,\n",
       "   'score': 54,\n",
       "   'last_activity_date': 1685874348,\n",
       "   'creation_date': 1650158620,\n",
       "   'question_id': 71898644,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/71898644/how-to-use-python-typing-annotated',\n",
       "   'title': 'How to use python typing.Annotated?'},\n",
       "  {'tags': ['python', 'python-poetry'],\n",
       "   'owner': {'account_id': 442058,\n",
       "    'reputation': 38,\n",
       "    'user_id': 832490,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 100,\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/38de998f10120ac8f0dab7f193bedf2b?s=256&d=identicon&r=PG&f=y&so-version=2',\n",
       "    'display_name': 'Rodrigo',\n",
       "    'link': 'https://stackoverflow.com/users/832490/rodrigo'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 34487,\n",
       "   'accepted_answer_id': 72547402,\n",
       "   'answer_count': 2,\n",
       "   'score': 54,\n",
       "   'last_activity_date': 1689952254,\n",
       "   'creation_date': 1652898582,\n",
       "   'last_edit_date': 1652904288,\n",
       "   'question_id': 72294299,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/72294299/multiple-top-level-packages-discovered-in-a-flat-layout',\n",
       "   'title': 'Multiple top-level packages discovered in a flat-layout'},\n",
       "  {'tags': ['python', 'openai-api'],\n",
       "   'owner': {'account_id': 19934883,\n",
       "    'reputation': 727,\n",
       "    'user_id': 14607802,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/ebb1a082fb58622f567e9f8ba1b480ca?s=256&d=identicon&r=PG&f=y&so-version=2',\n",
       "    'display_name': 'Charmalade',\n",
       "    'link': 'https://stackoverflow.com/users/14607802/charmalade'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 34589,\n",
       "   'answer_count': 7,\n",
       "   'score': 54,\n",
       "   'last_activity_date': 1696492784,\n",
       "   'creation_date': 1670380288,\n",
       "   'last_edit_date': 1683994435,\n",
       "   'question_id': 74711107,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/74711107/openai-api-continuing-conversation-in-a-dialogue',\n",
       "   'title': 'OpenAI API continuing conversation in a dialogue'},\n",
       "  {'tags': ['python'],\n",
       "   'owner': {'account_id': 24835532,\n",
       "    'reputation': 501,\n",
       "    'user_id': 18714220,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://lh3.googleusercontent.com/a-/AOh14GjpNxl2WQc7w30aQ9y641atIyL2lvv3QmWJaerZC04=k-s256',\n",
       "    'display_name': 'Bachir bessei',\n",
       "    'link': 'https://stackoverflow.com/users/18714220/bachir-bessei'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 120989,\n",
       "   'answer_count': 17,\n",
       "   'score': 50,\n",
       "   'last_activity_date': 1694893800,\n",
       "   'creation_date': 1649198983,\n",
       "   'last_edit_date': 1673098786,\n",
       "   'question_id': 71759248,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/71759248/importerror-cannot-import-name-builder-from-google-protobuf-internal',\n",
       "   'title': 'ImportError: cannot import name &#39;builder&#39; from &#39;google.protobuf.internal&#39;'},\n",
       "  {'tags': ['python', 'keras', 'python-import'],\n",
       "   'owner': {'account_id': 21037901,\n",
       "    'reputation': 500,\n",
       "    'user_id': 15460921,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://lh3.googleusercontent.com/a-/AOh14Gi7Gj0kCeAAyfShKeD_hYFIzMQnalEKsPLX3-NMFw=k-s256',\n",
       "    'display_name': 'adhi',\n",
       "    'link': 'https://stackoverflow.com/users/15460921/adhi'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 104764,\n",
       "   'protected_date': 1679307851,\n",
       "   'answer_count': 7,\n",
       "   'score': 50,\n",
       "   'last_activity_date': 1681697892,\n",
       "   'creation_date': 1653096480,\n",
       "   'last_edit_date': 1653153941,\n",
       "   'question_id': 72326025,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/72326025/cannot-import-name-pad-sequences-from-keras-preprocessing-sequence',\n",
       "   'title': 'cannot import name &#39;pad_sequences&#39; from &#39;keras.preprocessing.sequence&#39;'}],\n",
       " 'has_more': False,\n",
       " 'quota_max': 300,\n",
       " 'quota_remaining': 185}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of questions with the tag \"python\" between 2022-01-01 00:00:00 and 2022-12-31 00:00:00 and a minimum score of 50: \n",
      "35\n"
     ]
    }
   ],
   "source": [
    "# Seulement 10 questions crées avec le tag python et un score > 50 (en presque un an)\n",
    "\n",
    "# Observons l'activité l'année dernière\n",
    "start_date = datetime.datetime(2022, 1, 1)\n",
    "end_date = datetime.datetime(2022, 12, 31)\n",
    "\n",
    "result = get_number_of_questions(tag, start_date, end_date)\n",
    "print(f'Number of questions with the tag \"{tag}\" between {start_date} and {end_date} and a minimum score of 50: \\n{result}')\n",
    "\n",
    "# Il faut du temps pour que le score atteigne 50, mais le traffic sur stack exchange a\n",
    "# peut-être aussi diminué depuis l'année dernière (arrivée de chatGPT, etc...)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'items': [{'tags': ['python', 'pip', 'packaging', 'pyproject.toml'],\n",
       "   'owner': {'account_id': 3400218,\n",
       "    'reputation': 29218,\n",
       "    'user_id': 3015186,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 50,\n",
       "    'profile_image': 'https://i.stack.imgur.com/KagUg.jpg?s=256&g=1',\n",
       "    'display_name': 'Niko Fohr',\n",
       "    'link': 'https://stackoverflow.com/users/3015186/niko-fohr'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 267658,\n",
       "   'accepted_answer_id': 66472800,\n",
       "   'answer_count': 5,\n",
       "   'score': 343,\n",
       "   'last_activity_date': 1677001110,\n",
       "   'creation_date': 1595181079,\n",
       "   'last_edit_date': 1614931841,\n",
       "   'question_id': 62983756,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/62983756/what-is-pyproject-toml-file-for',\n",
       "   'title': 'What is pyproject.toml file for?'},\n",
       "  {'tags': ['python', 'pandas', 'xlrd', 'pcf'],\n",
       "   'owner': {'account_id': 20190357,\n",
       "    'reputation': 2889,\n",
       "    'user_id': 14808721,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://lh4.googleusercontent.com/-M5mYg2mtURw/AAAAAAAAAAI/AAAAAAAAAAA/AMZuuck8OiIkLBi6IBucITy4XIy72uK80A/s96-c/photo.jpg?sz=256',\n",
       "    'display_name': 'Vignesh K',\n",
       "    'link': 'https://stackoverflow.com/users/14808721/vignesh-k'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 577752,\n",
       "   'closed_date': 1608122436,\n",
       "   'answer_count': 2,\n",
       "   'score': 287,\n",
       "   'last_activity_date': 1612795847,\n",
       "   'creation_date': 1607702016,\n",
       "   'last_edit_date': 1612795847,\n",
       "   'question_id': 65254535,\n",
       "   'link': 'https://stackoverflow.com/questions/65254535/xlrd-biffh-xlrderror-excel-xlsx-file-not-supported',\n",
       "   'closed_reason': 'Duplicate',\n",
       "   'title': 'xlrd.biffh.XLRDError: Excel xlsx file; not supported'},\n",
       "  {'tags': ['python', 'machine-learning', 'deep-learning', 'pytorch'],\n",
       "   'owner': {'account_id': 543843,\n",
       "    'reputation': 24336,\n",
       "    'user_id': 913098,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 50,\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/e3e07b09128f889209b738276d42d7ef?s=256&d=identicon&r=PG',\n",
       "    'display_name': 'Gulzar',\n",
       "    'link': 'https://stackoverflow.com/users/913098/gulzar'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 263078,\n",
       "   'accepted_answer_id': 60018731,\n",
       "   'answer_count': 4,\n",
       "   'score': 259,\n",
       "   'last_activity_date': 1674767993,\n",
       "   'creation_date': 1580572695,\n",
       "   'last_edit_date': 1648471214,\n",
       "   'question_id': 60018578,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/60018578/what-does-model-eval-do-in-pytorch',\n",
       "   'title': 'What does model.eval() do in pytorch?'},\n",
       "  {'tags': ['python', 'pandas', 'numpy', 'scikit-learn', 'python-3.7'],\n",
       "   'owner': {'account_id': 6886384,\n",
       "    'reputation': 2351,\n",
       "    'user_id': 15150567,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://lh6.googleusercontent.com/-OWGodLwptYQ/AAAAAAAAAAI/AAAAAAAAA1o/AMZuucmZZ7jUgnhUF3BKhul3RbeBZAF1Bw/s96-c/photo.jpg?sz=256',\n",
       "    'display_name': 'Sachit Jani',\n",
       "    'link': 'https://stackoverflow.com/users/15150567/sachit-jani'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 330236,\n",
       "   'protected_date': 1673901248,\n",
       "   'answer_count': 19,\n",
       "   'score': 225,\n",
       "   'last_activity_date': 1680884751,\n",
       "   'creation_date': 1612516308,\n",
       "   'last_edit_date': 1673850793,\n",
       "   'question_id': 66060487,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/66060487/valueerror-numpy-ndarray-size-changed-may-indicate-binary-incompatibility-exp',\n",
       "   'title': 'ValueError: numpy.ndarray size changed, may indicate binary incompatibility. Expected 88 from C header, got 80 from PyObject'},\n",
       "  {'tags': ['python', 'linux', 'docker', 'docker-compose'],\n",
       "   'owner': {'account_id': 20032870,\n",
       "    'reputation': 2211,\n",
       "    'user_id': 14685247,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/0a20904979404a4f79e1d4f8bb405aab?s=256&d=identicon&r=PG&f=y&so-version=2',\n",
       "    'display_name': 'ming',\n",
       "    'link': 'https://stackoverflow.com/users/14685247/ming'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 333928,\n",
       "   'answer_count': 22,\n",
       "   'score': 221,\n",
       "   'last_activity_date': 1699410067,\n",
       "   'creation_date': 1606035195,\n",
       "   'last_edit_date': 1675820908,\n",
       "   'question_id': 64952238,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/64952238/docker-errors-dockerexception-error-while-fetching-server-api-version',\n",
       "   'title': 'docker.errors.DockerException: Error while fetching server API version'},\n",
       "  {'tags': ['python',\n",
       "    'postgresql',\n",
       "    'sqlalchemy',\n",
       "    'flask-sqlalchemy',\n",
       "    'psycopg2'],\n",
       "   'owner': {'account_id': 8971560,\n",
       "    'reputation': 2201,\n",
       "    'user_id': 6689838,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://graph.facebook.com/1072816459461942/picture?type=large',\n",
       "    'display_name': 'Maharsh Gheewala',\n",
       "    'link': 'https://stackoverflow.com/users/6689838/maharsh-gheewala'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 160916,\n",
       "   'protected_date': 1675825029,\n",
       "   'answer_count': 8,\n",
       "   'score': 220,\n",
       "   'last_activity_date': 1680670127,\n",
       "   'creation_date': 1593658606,\n",
       "   'last_edit_date': 1616648738,\n",
       "   'question_id': 62688256,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/62688256/sqlalchemy-exc-nosuchmoduleerror-cant-load-plugin-sqlalchemy-dialectspostgre',\n",
       "   'title': 'sqlalchemy.exc.NoSuchModuleError: Can&#39;t load plugin: sqlalchemy.dialects:postgres'},\n",
       "  {'tags': ['python', 'python-3.x', 'windows-10'],\n",
       "   'owner': {'account_id': 20238494,\n",
       "    'reputation': 2033,\n",
       "    'user_id': 14842976,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://lh4.googleusercontent.com/-tPgUuYJN48Q/AAAAAAAAAAI/AAAAAAAAAAA/AMZuuckXs8Wy0ReqrKZ98hL19LxSjnU7mw/s96-c/photo.jpg?sz=256',\n",
       "    'display_name': 'Joseph Haywood',\n",
       "    'link': 'https://stackoverflow.com/users/14842976/joseph-haywood'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 1231309,\n",
       "   'protected_date': 1615207834,\n",
       "   'closed_date': 1691903466,\n",
       "   'accepted_answer_id': 65349178,\n",
       "   'answer_count': 25,\n",
       "   'score': 203,\n",
       "   'last_activity_date': 1695144539,\n",
       "   'creation_date': 1608242137,\n",
       "   'last_edit_date': 1661715212,\n",
       "   'question_id': 65348890,\n",
       "   'link': 'https://stackoverflow.com/questions/65348890/python-was-not-found-run-without-arguments-to-install-from-the-microsoft-store',\n",
       "   'closed_reason': 'Duplicate',\n",
       "   'title': 'Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings'},\n",
       "  {'tags': ['python', 'selenium'],\n",
       "   'owner': {'account_id': 10383148,\n",
       "    'reputation': 9210,\n",
       "    'user_id': 7657180,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/6c0ee64d62bb9980080e6881a20a15fc?s=256&d=identicon&r=PG',\n",
       "    'display_name': 'YasserKhalil',\n",
       "    'link': 'https://stackoverflow.com/users/7657180/yasserkhalil'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 441207,\n",
       "   'protected_date': 1657217293,\n",
       "   'accepted_answer_id': 70099102,\n",
       "   'answer_count': 11,\n",
       "   'score': 198,\n",
       "   'last_activity_date': 1699941869,\n",
       "   'creation_date': 1604676159,\n",
       "   'question_id': 64717302,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/64717302/deprecationwarning-executable-path-has-been-deprecated-selenium-python',\n",
       "   'title': 'DeprecationWarning: executable_path has been deprecated selenium python'},\n",
       "  {'tags': ['python', 'macos', 'terminal', 'atom-editor', 'macos-monterey'],\n",
       "   'owner': {'account_id': 24663681,\n",
       "    'reputation': 3226,\n",
       "    'user_id': 18559642,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/13cdab4eb9d9fb5b1c7dece549f96046?s=256&d=identicon&r=PG',\n",
       "    'display_name': 'Kirill Ignatyev',\n",
       "    'link': 'https://stackoverflow.com/users/18559642/kirill-ignatyev'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 370728,\n",
       "   'protected_date': 1673118234,\n",
       "   'accepted_answer_id': 71621142,\n",
       "   'answer_count': 18,\n",
       "   'score': 197,\n",
       "   'last_activity_date': 1675219189,\n",
       "   'creation_date': 1648058522,\n",
       "   'last_edit_date': 1656277606,\n",
       "   'question_id': 71591971,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/71591971/how-can-i-fix-the-zsh-command-not-found-python-error-macos-monterey-12-3',\n",
       "   'title': 'How can I fix the &quot;zsh: command not found: python&quot; error? (macOS Monterey 12.3, Python&amp;nbsp;3.10, Atom IDE, and atom-python-run 0.9.7)'},\n",
       "  {'tags': ['python', 'scipy', 'pip', 'embedded-linux', 'python-wheel'],\n",
       "   'owner': {'account_id': 18141135,\n",
       "    'reputation': 2186,\n",
       "    'user_id': 13194982,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://lh4.googleusercontent.com/-VAtJtMr3kYU/AAAAAAAAAAI/AAAAAAAAAAA/AAKWJJMjwnVVurgGF-5_cRAsdHZDLzCo4Q/photo.jpg?sz=256',\n",
       "    'display_name': 'Sourabrt',\n",
       "    'link': 'https://stackoverflow.com/users/13194982/sourabrt'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 447023,\n",
       "   'answer_count': 30,\n",
       "   'score': 196,\n",
       "   'last_activity_date': 1658308129,\n",
       "   'creation_date': 1587560390,\n",
       "   'question_id': 61365790,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/61365790/error-could-not-build-wheels-for-scipy-which-use-pep-517-and-cannot-be-installe',\n",
       "   'title': 'ERROR: Could not build wheels for scipy which use PEP 517 and cannot be installed directly'},\n",
       "  {'tags': ['python', 'tensorflow', 'ray'],\n",
       "   'owner': {'account_id': 22989345,\n",
       "    'reputation': 1981,\n",
       "    'user_id': 17115086,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/16c137f7793988efb97ab410f7aec4f6?s=256&d=identicon&r=PG',\n",
       "    'display_name': 'hasu33',\n",
       "    'link': 'https://stackoverflow.com/users/17115086/hasu33'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 274176,\n",
       "   'protected_date': 1689017298,\n",
       "   'answer_count': 16,\n",
       "   'score': 193,\n",
       "   'last_activity_date': 1698317855,\n",
       "   'creation_date': 1653965255,\n",
       "   'last_edit_date': 1656248470,\n",
       "   'question_id': 72441758,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/72441758/typeerror-descriptors-cannot-not-be-created-directly',\n",
       "   'title': 'TypeError: Descriptors cannot not be created directly'},\n",
       "  {'tags': ['python', 'macos', 'homebrew', 'python-2.x'],\n",
       "   'owner': {'account_id': 3023753,\n",
       "    'reputation': 3616,\n",
       "    'user_id': 2564199,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 74,\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/e7de6198991a53a6c0da7ef3f5011725?s=256&d=identicon&r=PG',\n",
       "    'display_name': 'Pauline',\n",
       "    'link': 'https://stackoverflow.com/users/2564199/pauline'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 213196,\n",
       "   'accepted_answer_id': 60345962,\n",
       "   'answer_count': 13,\n",
       "   'score': 191,\n",
       "   'last_activity_date': 1675891874,\n",
       "   'creation_date': 1582109114,\n",
       "   'last_edit_date': 1602553702,\n",
       "   'question_id': 60298514,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/60298514/how-to-reinstall-python2-from-homebrew',\n",
       "   'title': 'How to reinstall python@2 from Homebrew?'},\n",
       "  {'tags': ['python', 'pip'],\n",
       "   'owner': {'account_id': 4071073,\n",
       "    'reputation': 4567,\n",
       "    'user_id': 3344139,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 79,\n",
       "    'profile_image': 'https://i.stack.imgur.com/ytB7I.jpg?s=256&g=1',\n",
       "    'display_name': 'RunOrVeith',\n",
       "    'link': 'https://stackoverflow.com/users/3344139/runorveith'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 102335,\n",
       "   'accepted_answer_id': 65793484,\n",
       "   'answer_count': 9,\n",
       "   'score': 176,\n",
       "   'last_activity_date': 1670330583,\n",
       "   'creation_date': 1606986952,\n",
       "   'last_edit_date': 1619177945,\n",
       "   'question_id': 65122957,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/65122957/resolving-new-pip-backtracking-runtime-issue',\n",
       "   'title': 'Resolving new pip backtracking runtime issue'},\n",
       "  {'tags': ['python', 'python-black'],\n",
       "   'owner': {'account_id': 13900281,\n",
       "    'reputation': 11703,\n",
       "    'user_id': 10036039,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/8790cbc891c02fffb597317b43a063fa?s=256&d=identicon&r=PG&f=y&so-version=2',\n",
       "    'display_name': 'zerocewl',\n",
       "    'link': 'https://stackoverflow.com/users/10036039/zerocewl'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 79903,\n",
       "   'accepted_answer_id': 71674345,\n",
       "   'answer_count': 6,\n",
       "   'score': 174,\n",
       "   'last_activity_date': 1679491903,\n",
       "   'creation_date': 1648626375,\n",
       "   'last_edit_date': 1651785651,\n",
       "   'question_id': 71673404,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/71673404/importerror-cannot-import-name-unicodefun-from-click',\n",
       "   'title': 'ImportError: cannot import name &#39;_unicodefun&#39; from &#39;click&#39;'},\n",
       "  {'tags': ['python', 'list', 'indexing', 'swap'],\n",
       "   'owner': {'account_id': 16490432,\n",
       "    'reputation': 2730,\n",
       "    'user_id': 13860719,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://i.stack.imgur.com/9aXXj.jpg?s=256&g=1',\n",
       "    'display_name': 'Shaun Han',\n",
       "    'link': 'https://stackoverflow.com/users/13860719/shaun-han'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 22682,\n",
       "   'accepted_answer_id': 68152940,\n",
       "   'answer_count': 8,\n",
       "   'score': 172,\n",
       "   'last_activity_date': 1698106750,\n",
       "   'creation_date': 1624808189,\n",
       "   'last_edit_date': 1698106750,\n",
       "   'question_id': 68152730,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/68152730/understand-python-swapping-why-is-a-b-b-a-not-always-equivalent-to-b-a-a',\n",
       "   'title': 'Understand Python swapping: why is a, b = b, a not always equivalent to b, a = a, b?'},\n",
       "  {'tags': ['python', 'python-3.x', 'django'],\n",
       "   'owner': {'account_id': 20471934,\n",
       "    'reputation': 1867,\n",
       "    'user_id': 15022111,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://lh3.googleusercontent.com/a-/AOh14GglbOwwii7QET_t9MBkZTDpGTbdvrOVCOV40SkWew=k-s256',\n",
       "    'display_name': 'Hi There',\n",
       "    'link': 'https://stackoverflow.com/users/15022111/hi-there'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 82973,\n",
       "   'accepted_answer_id': 66971803,\n",
       "   'answer_count': 5,\n",
       "   'score': 171,\n",
       "   'last_activity_date': 1659887345,\n",
       "   'creation_date': 1617722621,\n",
       "   'last_edit_date': 1659887345,\n",
       "   'question_id': 66971594,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/66971594/auto-create-primary-key-used-when-not-defining-a-primary-key-type-warning-in-dja',\n",
       "   'title': 'Auto-create primary key used when not defining a primary key type warning in Django'},\n",
       "  {'tags': ['python', 'excel', 'pandas'],\n",
       "   'owner': {'account_id': 20188429,\n",
       "    'reputation': 2607,\n",
       "    'user_id': 14807216,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/b31b40bb7f63928b9dc755251aa27fc1?s=256&d=identicon&r=PG&f=y&so-version=2',\n",
       "    'display_name': 'LNQ',\n",
       "    'link': 'https://stackoverflow.com/users/14807216/lnq'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 184591,\n",
       "   'protected_date': 1608777720,\n",
       "   'answer_count': 5,\n",
       "   'score': 171,\n",
       "   'last_activity_date': 1667945202,\n",
       "   'creation_date': 1607684916,\n",
       "   'last_edit_date': 1607938283,\n",
       "   'question_id': 65250207,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/65250207/pandas-cannot-open-an-excel-xlsx-file',\n",
       "   'title': 'Pandas cannot open an Excel (.xlsx) file'},\n",
       "  {'tags': ['python', 'django', 'django-urls', 'django-4.0'],\n",
       "   'owner': {'account_id': 39506,\n",
       "    'reputation': 301205,\n",
       "    'user_id': 113962,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 87,\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/b80ae1d65878cf68a4b1a00848467527?s=256&d=identicon&r=PG',\n",
       "    'display_name': 'Alasdair',\n",
       "    'link': 'https://stackoverflow.com/users/113962/alasdair'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 189803,\n",
       "   'protected_date': 1688169134,\n",
       "   'accepted_answer_id': 70319607,\n",
       "   'answer_count': 5,\n",
       "   'score': 169,\n",
       "   'last_activity_date': 1690610954,\n",
       "   'creation_date': 1639262655,\n",
       "   'question_id': 70319606,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/70319606/importerror-cannot-import-name-url-from-django-conf-urls-after-upgrading-to',\n",
       "   'title': 'ImportError: cannot import name &#39;url&#39; from &#39;django.conf.urls&#39; after upgrading to Django 4.0'},\n",
       "  {'tags': ['python', 'python-3.x', 'tensorflow', 'keras', 'tensorflow2.0'],\n",
       "   'owner': {'account_id': 3882738,\n",
       "    'reputation': 18649,\n",
       "    'user_id': 3214872,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 82,\n",
       "    'profile_image': 'https://i.stack.imgur.com/NJBvG.jpg?s=256&g=1',\n",
       "    'display_name': 'GPhilo',\n",
       "    'link': 'https://stackoverflow.com/users/3214872/gphilo'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 463485,\n",
       "   'protected_date': 1596965559,\n",
       "   'accepted_answer_id': 59823284,\n",
       "   'answer_count': 19,\n",
       "   'score': 160,\n",
       "   'last_activity_date': 1667973558,\n",
       "   'creation_date': 1579523213,\n",
       "   'last_edit_date': 1590593147,\n",
       "   'question_id': 59823283,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/59823283/could-not-load-dynamic-library-cudart64-101-dll-on-tensorflow-cpu-only-install',\n",
       "   'title': 'Could not load dynamic library &#39;cudart64_101.dll&#39; on tensorflow CPU-only installation'},\n",
       "  {'tags': ['python', 'linux', 'alpine-linux'],\n",
       "   'owner': {'account_id': 6547232,\n",
       "    'reputation': 2129,\n",
       "    'user_id': 5062828,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 22,\n",
       "    'profile_image': 'https://lh5.googleusercontent.com/-76-PjkdknXw/AAAAAAAAAAI/AAAAAAAAAAA/ACevoQPo1bNLcNrv-ki3G8Lnlad_JmAClQ/mo/photo.jpg?sz=256',\n",
       "    'display_name': 'Cutaraca',\n",
       "    'link': 'https://stackoverflow.com/users/5062828/cutaraca'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 275509,\n",
       "   'accepted_answer_id': 62555259,\n",
       "   'answer_count': 8,\n",
       "   'score': 155,\n",
       "   'last_activity_date': 1687675164,\n",
       "   'creation_date': 1593001526,\n",
       "   'question_id': 62554991,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/62554991/how-do-i-install-python-on-alpine-linux',\n",
       "   'title': 'How do I install python on alpine linux?'},\n",
       "  {'tags': ['python', 'error-handling', 'pip'],\n",
       "   'owner': {'account_id': 23480036,\n",
       "    'reputation': 1614,\n",
       "    'user_id': 17532318,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/d37672946ecec10b2814be93daa910ef?s=256&d=identicon&r=PG',\n",
       "    'display_name': 'Apoliticalboy',\n",
       "    'link': 'https://stackoverflow.com/users/17532318/apoliticalboy'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 224796,\n",
       "   'answer_count': 17,\n",
       "   'score': 150,\n",
       "   'last_activity_date': 1700813401,\n",
       "   'creation_date': 1677700339,\n",
       "   'last_edit_date': 1694084649,\n",
       "   'question_id': 75608323,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/75608323/how-do-i-solve-error-externally-managed-environment-every-time-i-use-pip-3',\n",
       "   'title': 'How do I solve &quot;error: externally-managed-environment&quot; every time I use pip 3?'},\n",
       "  {'tags': ['python', 'pandas', 'dataframe', 'attributeerror'],\n",
       "   'owner': {'account_id': 18651684,\n",
       "    'reputation': 1673,\n",
       "    'user_id': 13595011,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://lh3.googleusercontent.com/a-/AOh14Gja1Bf3zfNP7LY4HSFgp6beG2pul2AVcXIsoLC52g=k-s256',\n",
       "    'display_name': 'Maksimjeet Chowdhary',\n",
       "    'link': 'https://stackoverflow.com/users/13595011/maksimjeet-chowdhary'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 261517,\n",
       "   'protected_date': 1695993982,\n",
       "   'answer_count': 3,\n",
       "   'score': 149,\n",
       "   'last_activity_date': 1699567288,\n",
       "   'creation_date': 1680851159,\n",
       "   'last_edit_date': 1691967377,\n",
       "   'question_id': 75956209,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/75956209/error-dataframe-object-has-no-attribute-append',\n",
       "   'title': 'Error &quot;&#39;DataFrame&#39; object has no attribute &#39;append&#39;&quot;'},\n",
       "  {'tags': ['python', 'python-3.x', 'list', 'cpython', 'python-internals'],\n",
       "   'owner': {'account_id': 1843521,\n",
       "    'reputation': 27723,\n",
       "    'user_id': 1672429,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 77,\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/0815eb7d4e9dd58022d51ba2011b9c0a?s=256&d=identicon&r=PG',\n",
       "    'display_name': 'Stefan Pochmann',\n",
       "    'link': 'https://stackoverflow.com/users/1672429/stefan-pochmann'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 6102,\n",
       "   'accepted_answer_id': 60551320,\n",
       "   'answer_count': 3,\n",
       "   'score': 149,\n",
       "   'last_activity_date': 1583600431,\n",
       "   'creation_date': 1583425357,\n",
       "   'last_edit_date': 1583429583,\n",
       "   'question_id': 60549865,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/60549865/what-causes-a-to-overallocate',\n",
       "   'title': 'What causes [*a] to overallocate?'},\n",
       "  {'tags': ['python', 'python-3.x', 'visual-studio'],\n",
       "   'owner': {'account_id': 19662853,\n",
       "    'reputation': 1420,\n",
       "    'user_id': 14393948,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://lh3.googleusercontent.com/a-/AOh14Gg43aRRTvCexfkiavPSoo5UpYnVZY_0K6AcxdI6bg=k-s256',\n",
       "    'display_name': 'wloleo',\n",
       "    'link': 'https://stackoverflow.com/users/14393948/wloleo'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 358694,\n",
       "   'protected_date': 1668477194,\n",
       "   'accepted_answer_id': 64262038,\n",
       "   'answer_count': 8,\n",
       "   'score': 147,\n",
       "   'last_activity_date': 1699328956,\n",
       "   'creation_date': 1602156410,\n",
       "   'last_edit_date': 1647686567,\n",
       "   'question_id': 64261546,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/64261546/how-to-solve-error-microsoft-visual-c-14-0-or-greater-is-required-when-inst',\n",
       "   'title': 'How to solve &quot;error: Microsoft Visual C++ 14.0 or greater is required&quot; when installing Python packages?'},\n",
       "  {'tags': ['python', 'visual-studio-code', 'python-poetry'],\n",
       "   'owner': {'account_id': 10908059,\n",
       "    'reputation': 3074,\n",
       "    'user_id': 8491363,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/6efa1647cc8d6adae97bcab9fd30d9ef?s=256&d=identicon&r=PG',\n",
       "    'display_name': 'user8491363',\n",
       "    'link': 'https://stackoverflow.com/users/8491363/user8491363'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 86444,\n",
       "   'accepted_answer_id': 64434542,\n",
       "   'answer_count': 4,\n",
       "   'score': 146,\n",
       "   'last_activity_date': 1695631489,\n",
       "   'creation_date': 1579796188,\n",
       "   'question_id': 59882884,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/59882884/vscode-doesnt-show-poetry-virtualenvs-in-select-interpreter-option',\n",
       "   'title': 'VSCode doesn&#39;t show poetry virtualenvs in select interpreter option'},\n",
       "  {'tags': ['python', 'path', 'pip', 'requirements.txt'],\n",
       "   'owner': {'account_id': 8311384,\n",
       "    'reputation': 2071,\n",
       "    'user_id': 6246352,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://i.stack.imgur.com/ws9ud.jpg?s=256&g=1',\n",
       "    'display_name': 'Naeem Khoshnevis',\n",
       "    'link': 'https://stackoverflow.com/users/6246352/naeem-khoshnevis'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 52488,\n",
       "   'protected_date': 1662599526,\n",
       "   'accepted_answer_id': 62886215,\n",
       "   'answer_count': 1,\n",
       "   'score': 144,\n",
       "   'last_activity_date': 1652418020,\n",
       "   'creation_date': 1594683418,\n",
       "   'question_id': 62885911,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/62885911/pip-freeze-creates-some-weird-path-instead-of-the-package-version',\n",
       "   'title': 'pip freeze creates some weird path instead of the package version'},\n",
       "  {'tags': ['python', 'pip', 'python-poetry'],\n",
       "   'owner': {'account_id': 446266,\n",
       "    'reputation': 6482,\n",
       "    'user_id': 839211,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 90,\n",
       "    'profile_image': 'https://i.stack.imgur.com/THWvr.jpg?s=256&g=1',\n",
       "    'display_name': 'Felipe',\n",
       "    'link': 'https://stackoverflow.com/users/839211/felipe'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 105583,\n",
       "   'accepted_answer_id': 62787881,\n",
       "   'answer_count': 16,\n",
       "   'score': 143,\n",
       "   'last_activity_date': 1700488456,\n",
       "   'creation_date': 1594067909,\n",
       "   'last_edit_date': 1662582741,\n",
       "   'question_id': 62764148,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/62764148/how-to-import-requirements-txt-from-an-existing-project-using-poetry',\n",
       "   'title': 'How to import requirements.txt from an existing project using Poetry'},\n",
       "  {'tags': ['python', 'machine-learning', 'package', 'conda', 'python-poetry'],\n",
       "   'owner': {'account_id': 1449869,\n",
       "    'reputation': 2561,\n",
       "    'user_id': 1367124,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 68,\n",
       "    'profile_image': 'https://i.stack.imgur.com/azGrD.jpg?s=256&g=1',\n",
       "    'display_name': 'Seub',\n",
       "    'link': 'https://stackoverflow.com/users/1367124/seub'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 61264,\n",
       "   'accepted_answer_id': 71110028,\n",
       "   'answer_count': 3,\n",
       "   'score': 142,\n",
       "   'last_activity_date': 1699380700,\n",
       "   'creation_date': 1643123383,\n",
       "   'last_edit_date': 1643126235,\n",
       "   'question_id': 70851048,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/70851048/does-it-make-sense-to-use-conda-poetry',\n",
       "   'title': 'Does it make sense to use Conda + Poetry?'},\n",
       "  {'tags': ['python', 'pandas', 'dataframe', 'data-wrangling', 'data-munging'],\n",
       "   'owner': {'account_id': 29727,\n",
       "    'reputation': 4413,\n",
       "    'user_id': 80544,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 79,\n",
       "    'profile_image': 'https://i.stack.imgur.com/HtN07.jpg?s=256&g=1',\n",
       "    'display_name': 'Glenn',\n",
       "    'link': 'https://stackoverflow.com/users/80544/glenn'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 175243,\n",
       "   'accepted_answer_id': 70837545,\n",
       "   'answer_count': 8,\n",
       "   'score': 142,\n",
       "   'last_activity_date': 1690552412,\n",
       "   'creation_date': 1643042804,\n",
       "   'last_edit_date': 1645482240,\n",
       "   'question_id': 70837397,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/70837397/good-alternative-to-pandas-append-method-now-that-it-is-being-deprecated',\n",
       "   'title': 'Good alternative to Pandas .append() method, now that it is being deprecated?'},\n",
       "  {'tags': ['python', 'python-3.x', 'windows', 'numpy'],\n",
       "   'owner': {'account_id': 19883764,\n",
       "    'reputation': 1403,\n",
       "    'user_id': 14567162,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://lh4.googleusercontent.com/-8ED7m_am4Uo/AAAAAAAAAAI/AAAAAAAAAbU/AMZuuclxX7dkrrBdmjNHqH7VsijjuTKGIg/s96-c/photo.jpg?sz=256',\n",
       "    'display_name': 'Reed Collins',\n",
       "    'link': 'https://stackoverflow.com/users/14567162/reed-collins'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 85465,\n",
       "   'protected_date': 1605219106,\n",
       "   'accepted_answer_id': 64658254,\n",
       "   'answer_count': 9,\n",
       "   'score': 140,\n",
       "   'last_activity_date': 1610096233,\n",
       "   'creation_date': 1604360224,\n",
       "   'last_edit_date': 1605519690,\n",
       "   'question_id': 64654805,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/64654805/how-do-you-fix-runtimeerror-package-fails-to-pass-a-sanity-check-for-numpy-an',\n",
       "   'title': 'How do you fix &quot;runtimeError: package fails to pass a sanity check&quot; for numpy and pandas?'},\n",
       "  {'tags': ['python', 'fastapi', 'uvicorn'],\n",
       "   'owner': {'account_id': 18038692,\n",
       "    'reputation': 1391,\n",
       "    'user_id': 13111448,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/b0668a7d54439dc6d0e453488bf53827?s=256&d=identicon&r=PG&f=y&so-version=2',\n",
       "    'display_name': 'Pokisutra',\n",
       "    'link': 'https://stackoverflow.com/users/13111448/pokisutra'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 203473,\n",
       "   'protected_date': 1668645866,\n",
       "   'accepted_answer_id': 62934660,\n",
       "   'answer_count': 14,\n",
       "   'score': 137,\n",
       "   'last_activity_date': 1693004003,\n",
       "   'creation_date': 1584987372,\n",
       "   'last_edit_date': 1648939338,\n",
       "   'question_id': 60819376,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/60819376/fastapi-throws-an-error-error-loading-asgi-app-could-not-import-module-api',\n",
       "   'title': 'FastAPI throws an error (Error loading ASGI app. Could not import module &quot;api&quot;)'},\n",
       "  {'tags': ['python', 'tensorflow', 'anaconda'],\n",
       "   'owner': {'account_id': 10181486,\n",
       "    'reputation': 3279,\n",
       "    'user_id': 7518091,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 91,\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/f8af9d177a327eaf22be1651863a103a?s=256&d=identicon&r=PG&f=y&so-version=2',\n",
       "    'display_name': 'user1315789',\n",
       "    'link': 'https://stackoverflow.com/users/7518091/user1315789'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 228875,\n",
       "   'accepted_answer_id': 65333085,\n",
       "   'answer_count': 8,\n",
       "   'score': 134,\n",
       "   'last_activity_date': 1700233427,\n",
       "   'creation_date': 1607990716,\n",
       "   'last_edit_date': 1700233279,\n",
       "   'question_id': 65298241,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/65298241/what-does-this-tensorflow-message-mean-any-side-effect-was-the-installation-su',\n",
       "   'title': 'What does this tensorflow message mean? Any side effect? Was the installation successful?'},\n",
       "  {'tags': ['python', 'pip'],\n",
       "   'owner': {'account_id': 12235881,\n",
       "    'reputation': 2037,\n",
       "    'user_id': 8930321,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 100,\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/0bc45c7114227787d7f27ce17f523a36?s=256&d=identicon&r=PG&f=y&so-version=2',\n",
       "    'display_name': 'lowercase00',\n",
       "    'link': 'https://stackoverflow.com/users/8930321/lowercase00'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 493124,\n",
       "   'protected_date': 1587068882,\n",
       "   'answer_count': 18,\n",
       "   'score': 132,\n",
       "   'last_activity_date': 1677439515,\n",
       "   'creation_date': 1580436978,\n",
       "   'last_edit_date': 1614469552,\n",
       "   'question_id': 59997065,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/59997065/pip-python-normal-site-packages-is-not-writeable',\n",
       "   'title': 'pip/python: normal site-packages is not writeable'},\n",
       "  {'tags': ['python', 'python-3.x', 'download', 'youtube', 'youtube-dl'],\n",
       "   'owner': {'account_id': 18939012,\n",
       "    'reputation': 1457,\n",
       "    'user_id': 13819476,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/8bbd0a28d4d0e1ce6ba2a58b63aae260?s=256&d=identicon&r=PG&f=y&so-version=2',\n",
       "    'display_name': 'Bastien',\n",
       "    'link': 'https://stackoverflow.com/users/13819476/bastien'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 175846,\n",
       "   'protected_date': 1609065407,\n",
       "   'accepted_answer_id': 65230976,\n",
       "   'answer_count': 10,\n",
       "   'score': 131,\n",
       "   'last_activity_date': 1682517250,\n",
       "   'creation_date': 1599673099,\n",
       "   'last_edit_date': 1619130962,\n",
       "   'question_id': 63816790,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/63816790/youtube-dl-error-youtube-said-unable-to-extract-video-data',\n",
       "   'title': 'Youtube_dl : ERROR : YouTube said: Unable to extract video data'},\n",
       "  {'tags': ['python', 'google-analytics-api'],\n",
       "   'owner': {'account_id': 6468208,\n",
       "    'reputation': 6455,\n",
       "    'user_id': 5009293,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 76,\n",
       "    'profile_image': 'https://graph.facebook.com/995762863775266/picture?type=large',\n",
       "    'display_name': 'Doug Coats',\n",
       "    'link': 'https://stackoverflow.com/users/5009293/doug-coats'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 205518,\n",
       "   'accepted_answer_id': 73932621,\n",
       "   'answer_count': 18,\n",
       "   'score': 128,\n",
       "   'last_activity_date': 1693903144,\n",
       "   'creation_date': 1663949925,\n",
       "   'last_edit_date': 1663952437,\n",
       "   'question_id': 73830524,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/73830524/attributeerror-module-lib-has-no-attribute-x509-v-flag-cb-issuer-check',\n",
       "   'title': 'AttributeError: module &#39;lib&#39; has no attribute &#39;X509_V_FLAG_CB_ISSUER_CHECK&#39;'},\n",
       "  {'tags': ['python', 'python-3.x', 'pathlib'],\n",
       "   'owner': {'account_id': 16410122,\n",
       "    'reputation': 1933,\n",
       "    'user_id': 11853632,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/12a934b35c313adcbf98aaf10aed69f1?s=256&d=identicon&r=PG&f=y&so-version=2',\n",
       "    'display_name': 'reinhardt',\n",
       "    'link': 'https://stackoverflow.com/users/11853632/reinhardt'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 94191,\n",
       "   'accepted_answer_id': 61321618,\n",
       "   'answer_count': 6,\n",
       "   'score': 123,\n",
       "   'last_activity_date': 1658545772,\n",
       "   'creation_date': 1587382490,\n",
       "   'question_id': 61321503,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/61321503/is-there-a-pathlib-alternate-for-os-path-join',\n",
       "   'title': 'Is there a Pathlib alternate for os.path.join?'},\n",
       "  {'tags': ['python', 'pylint'],\n",
       "   'owner': {'account_id': 15406446,\n",
       "    'reputation': 1419,\n",
       "    'user_id': 12160503,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://graph.facebook.com/2048851171863469/picture?type=large',\n",
       "    'display_name': 'Isaac Anatolio',\n",
       "    'link': 'https://stackoverflow.com/users/12160503/isaac-anatolio'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 193654,\n",
       "   'closed_date': 1665543271,\n",
       "   'answer_count': 8,\n",
       "   'score': 123,\n",
       "   'last_activity_date': 1628071953,\n",
       "   'creation_date': 1583716399,\n",
       "   'last_edit_date': 1616202315,\n",
       "   'question_id': 60593604,\n",
       "   'link': 'https://stackoverflow.com/questions/60593604/importerror-attempted-relative-import-with-no-known-parent-package',\n",
       "   'closed_reason': 'Duplicate',\n",
       "   'title': 'ImportError : Attempted relative import with no known parent package'},\n",
       "  {'tags': ['python',\n",
       "    'django',\n",
       "    'importerror',\n",
       "    'python-module',\n",
       "    'circular-dependency'],\n",
       "   'owner': {'account_id': 9473820,\n",
       "    'reputation': 1215,\n",
       "    'user_id': 7043835,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/4203ed009a7d8e7636fcfe3c0e03ae22?s=256&d=identicon&r=PG&f=y&so-version=2',\n",
       "    'display_name': '3WZ',\n",
       "    'link': 'https://stackoverflow.com/users/7043835/3wz'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 380536,\n",
       "   'protected_date': 1623706069,\n",
       "   'accepted_answer_id': 64809784,\n",
       "   'answer_count': 16,\n",
       "   'score': 121,\n",
       "   'last_activity_date': 1688392928,\n",
       "   'creation_date': 1605196086,\n",
       "   'last_edit_date': 1669973149,\n",
       "   'question_id': 64807163,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/64807163/importerror-cannot-import-name-from-partially-initialized-module-m',\n",
       "   'title': 'ImportError: cannot import name &#39;...&#39; from partially initialized module &#39;...&#39; (most likely due to a circular import)'},\n",
       "  {'tags': ['python',\n",
       "    'docker',\n",
       "    'docker-compose',\n",
       "    'dependencies',\n",
       "    'importerror'],\n",
       "   'owner': {'account_id': 3369545,\n",
       "    'reputation': 3171,\n",
       "    'user_id': 2829150,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 51,\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/1e9c788890b3d2a8e468ed2773864b2c?s=256&d=identicon&r=PG&f=y&so-version=2',\n",
       "    'display_name': 'Arie',\n",
       "    'link': 'https://stackoverflow.com/users/2829150/arie'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 152032,\n",
       "   'protected_date': 1691442712,\n",
       "   'answer_count': 15,\n",
       "   'score': 117,\n",
       "   'last_activity_date': 1699634829,\n",
       "   'creation_date': 1652208276,\n",
       "   'last_edit_date': 1667754370,\n",
       "   'question_id': 72191560,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/72191560/importerror-cannot-import-name-soft-unicode-from-markupsafe',\n",
       "   'title': 'ImportError: cannot import name &#39;soft_unicode&#39; from &#39;markupsafe&#39;'},\n",
       "  {'tags': ['python', 'pandas', 'dataframe'],\n",
       "   'owner': {'account_id': 15702272,\n",
       "    'reputation': 1261,\n",
       "    'user_id': 11330069,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://lh3.googleusercontent.com/-V1jQ5QLfKI8/AAAAAAAAAAI/AAAAAAAAABY/j4kK5yAK4Ak/photo.jpg?sz=256',\n",
       "    'display_name': 'Diarmid Roberts',\n",
       "    'link': 'https://stackoverflow.com/users/11330069/diarmid-roberts'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 161464,\n",
       "   'answer_count': 16,\n",
       "   'score': 116,\n",
       "   'last_activity_date': 1685493073,\n",
       "   'creation_date': 1580405248,\n",
       "   'last_edit_date': 1614713321,\n",
       "   'question_id': 59991397,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/59991397/attributeerror-dataframe-object-has-no-attribute-ix',\n",
       "   'title': 'AttributeError: &#39;DataFrame&#39; object has no attribute &#39;ix&#39;'},\n",
       "  {'tags': ['python', 'python-3.x', 'linux', 'pip'],\n",
       "   'owner': {'account_id': 22984508,\n",
       "    'reputation': 1161,\n",
       "    'user_id': 17110906,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/4d835ba2d974f7b6e2c5844a29641106?s=256&d=identicon&r=PG',\n",
       "    'display_name': 'Christopher Coco',\n",
       "    'link': 'https://stackoverflow.com/users/17110906/christopher-coco'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 142053,\n",
       "   'answer_count': 11,\n",
       "   'score': 115,\n",
       "   'last_activity_date': 1676514324,\n",
       "   'creation_date': 1633747154,\n",
       "   'last_edit_date': 1633747387,\n",
       "   'question_id': 69503329,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/69503329/pip-is-not-working-for-python-3-10-on-ubuntu',\n",
       "   'title': 'Pip is not working for Python 3.10 on Ubuntu'},\n",
       "  {'tags': ['python',\n",
       "    'machine-learning',\n",
       "    'scikit-learn',\n",
       "    'logistic-regression'],\n",
       "   'owner': {'account_id': 1130688,\n",
       "    'reputation': 1813,\n",
       "    'user_id': 7373787,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 89,\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/6df86729130c9a20a2df0e2aeefd7443?s=256&d=identicon&r=PG&f=y&so-version=2',\n",
       "    'display_name': 'sums22',\n",
       "    'link': 'https://stackoverflow.com/users/7373787/sums22'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 239718,\n",
       "   'protected_date': 1618936399,\n",
       "   'accepted_answer_id': 62659927,\n",
       "   'answer_count': 4,\n",
       "   'score': 115,\n",
       "   'last_activity_date': 1673518122,\n",
       "   'creation_date': 1593522500,\n",
       "   'last_edit_date': 1618913680,\n",
       "   'question_id': 62658215,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/62658215/convergencewarning-lbfgs-failed-to-converge-status-1-stop-total-no-of-iter',\n",
       "   'title': 'ConvergenceWarning: lbfgs failed to converge (status=1): STOP: TOTAL NO. of ITERATIONS REACHED LIMIT'},\n",
       "  {'tags': ['python', 'python-3.9'],\n",
       "   'owner': {'account_id': 4650123,\n",
       "    'reputation': 34595,\n",
       "    'user_id': 3767239,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 64,\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/4d6c9d978220f4ac99f6a1237288a99c?s=256&d=identicon&r=PG&f=y&so-version=2',\n",
       "    'display_name': 'a_guest',\n",
       "    'link': 'https://stackoverflow.com/users/3767239/a-guest'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 11147,\n",
       "   'answer_count': 3,\n",
       "   'score': 114,\n",
       "   'last_activity_date': 1612271090,\n",
       "   'creation_date': 1611094722,\n",
       "   'last_edit_date': 1611234002,\n",
       "   'question_id': 65800344,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/65800344/what-are-soft-keywords',\n",
       "   'title': 'What are &quot;soft keywords&quot;?'},\n",
       "  {'tags': ['python', 'pip'],\n",
       "   'owner': {'account_id': 69666,\n",
       "    'reputation': 6265,\n",
       "    'user_id': 202576,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 47,\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/bdf45a8317ff3471cdcc7ba2f18c79a1?s=256&d=identicon&r=PG',\n",
       "    'display_name': 'onknows',\n",
       "    'link': 'https://stackoverflow.com/users/202576/onknows'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 151719,\n",
       "   'protected_date': 1618889077,\n",
       "   'answer_count': 12,\n",
       "   'score': 113,\n",
       "   'last_activity_date': 1674357448,\n",
       "   'creation_date': 1611639213,\n",
       "   'last_edit_date': 1674357448,\n",
       "   'question_id': 65896334,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/65896334/upgrading-pip-fails-with-syntax-error-caused-by-sys-stderr-writeferror-exc',\n",
       "   'title': 'Upgrading pip fails with syntax error caused by sys.stderr.write(f&quot;ERROR: {exc}&quot;)'},\n",
       "  {'tags': ['python', 'virtualenv'],\n",
       "   'owner': {'account_id': 15854936,\n",
       "    'reputation': 1251,\n",
       "    'user_id': 11439612,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://i.stack.imgur.com/bBSil.jpg?s=256&g=1',\n",
       "    'display_name': 'Tung Ng.',\n",
       "    'link': 'https://stackoverflow.com/users/11439612/tung-ng'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 98039,\n",
       "   'accepted_answer_id': 63491349,\n",
       "   'answer_count': 15,\n",
       "   'score': 111,\n",
       "   'last_activity_date': 1694164402,\n",
       "   'creation_date': 1597854368,\n",
       "   'last_edit_date': 1676018186,\n",
       "   'question_id': 63491221,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/63491221/modulenotfounderror-no-module-named-virtualenv-seed-embed-via-app-data-when-i',\n",
       "   'title': 'ModuleNotFoundError: No module named &#39;virtualenv.seed.embed.via_app_data&#39; when I created new env by virtualenv'},\n",
       "  {'tags': ['python', 'numpy'],\n",
       "   'owner': {'account_id': 8369136,\n",
       "    'reputation': 3442,\n",
       "    'user_id': 6284785,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 50,\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/9d2ef82797b12ebbeebbf008f9524f50?s=256&d=identicon&r=PG',\n",
       "    'display_name': 'Anatoly',\n",
       "    'link': 'https://stackoverflow.com/users/6284785/anatoly'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 213919,\n",
       "   'answer_count': 22,\n",
       "   'score': 110,\n",
       "   'last_activity_date': 1695172833,\n",
       "   'creation_date': 1614021415,\n",
       "   'last_edit_date': 1614030641,\n",
       "   'question_id': 66322049,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/66322049/could-not-install-packages-due-to-an-oserror-winerror-2-no-such-file-or-direc',\n",
       "   'title': 'Could not install packages due to an OSError: [WinError 2] No such file or directory'},\n",
       "  {'tags': ['python', 'ubuntu', 'ubuntu-20.04'],\n",
       "   'owner': {'account_id': 13483168,\n",
       "    'reputation': 2689,\n",
       "    'user_id': 9727793,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://lh3.googleusercontent.com/-XdUIqdMkCWA/AAAAAAAAAAI/AAAAAAAAAAA/4252rscbv5M/photo.jpg?sz=256',\n",
       "    'display_name': 'Kh4zit',\n",
       "    'link': 'https://stackoverflow.com/users/9727793/kh4zit'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 108356,\n",
       "   'accepted_answer_id': 61876234,\n",
       "   'answer_count': 9,\n",
       "   'score': 110,\n",
       "   'last_activity_date': 1691484356,\n",
       "   'creation_date': 1589824360,\n",
       "   'last_edit_date': 1596652896,\n",
       "   'question_id': 61875869,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/61875869/ubuntu-20-04-upgrade-python-missing-libffi-so-6',\n",
       "   'title': 'Ubuntu 20.04 upgrade, Python missing libffi.so.6'},\n",
       "  {'tags': ['python',\n",
       "    'selenium',\n",
       "    'google-chrome',\n",
       "    'selenium-webdriver',\n",
       "    'selenium-chromedriver'],\n",
       "   'owner': {'account_id': 19488994,\n",
       "    'reputation': 1083,\n",
       "    'user_id': 14257612,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://www.gravatar.com/avatar/c49440f3d6f3831211bcb56d9614a61e?s=256&d=identicon&r=PG&f=y&so-version=2',\n",
       "    'display_name': 'Alesss',\n",
       "    'link': 'https://stackoverflow.com/users/14257612/alesss'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 251439,\n",
       "   'accepted_answer_id': 72773269,\n",
       "   'answer_count': 4,\n",
       "   'score': 108,\n",
       "   'last_activity_date': 1700498304,\n",
       "   'creation_date': 1656337711,\n",
       "   'last_edit_date': 1661689426,\n",
       "   'question_id': 72773206,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/72773206/selenium-python-attributeerror-webdriver-object-has-no-attribute-find-el',\n",
       "   'title': 'Selenium - Python - AttributeError: &#39;WebDriver&#39; object has no attribute &#39;find_element_by_name&#39;'},\n",
       "  {'tags': ['python', 'linux', 'openssl', 'python-poetry'],\n",
       "   'owner': {'account_id': 7489950,\n",
       "    'reputation': 4646,\n",
       "    'user_id': 5692012,\n",
       "    'user_type': 'registered',\n",
       "    'accept_rate': 50,\n",
       "    'profile_image': 'https://i.stack.imgur.com/a3FwR.gif?s=256&g=1',\n",
       "    'display_name': 'RobinFrcd',\n",
       "    'link': 'https://stackoverflow.com/users/5692012/robinfrcd'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 193080,\n",
       "   'accepted_answer_id': 72633324,\n",
       "   'answer_count': 14,\n",
       "   'score': 107,\n",
       "   'last_activity_date': 1699036115,\n",
       "   'creation_date': 1651782719,\n",
       "   'last_edit_date': 1675089528,\n",
       "   'question_id': 72133316,\n",
       "   'content_license': 'CC BY-SA 4.0',\n",
       "   'link': 'https://stackoverflow.com/questions/72133316/libssl-so-1-1-cannot-open-shared-object-file-no-such-file-or-directory',\n",
       "   'title': 'libssl.so.1.1: cannot open shared object file: No such file or directory'},\n",
       "  {'tags': ['python', 'python-3.x'],\n",
       "   'owner': {'account_id': 12560977,\n",
       "    'reputation': 1353,\n",
       "    'user_id': 10755032,\n",
       "    'user_type': 'registered',\n",
       "    'profile_image': 'https://lh4.googleusercontent.com/-OCvviTeYfqU/AAAAAAAAAAI/AAAAAAAAAA0/94yUImsA6hc/photo.jpg?sz=256',\n",
       "    'display_name': 'Karthik Bhandary',\n",
       "    'link': 'https://stackoverflow.com/users/10755032/karthik-bhandary'},\n",
       "   'is_answered': True,\n",
       "   'view_count': 256264,\n",
       "   'protected_date': 1607426163,\n",
       "   'closed_date': 1663684177,\n",
       "   'accepted_answer_id': 59777118,\n",
       "   'answer_count': 4,\n",
       "   'score': 107,\n",
       "   'last_activity_date': 1681200762,\n",
       "   'creation_date': 1579148521,\n",
       "   'last_edit_date': 1609769398,\n",
       "   'question_id': 59762996,\n",
       "   'link': 'https://stackoverflow.com/questions/59762996/how-to-fix-attributeerror-partially-initialized-module',\n",
       "   'closed_reason': 'Duplicate',\n",
       "   'title': 'How to fix AttributeError: partially initialized module?'}],\n",
       " 'has_more': True,\n",
       " 'quota_max': 300,\n",
       " 'quota_remaining': 184}"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "def get_questions(tag, start_date, end_date, min_score=50):\n",
    "    api_url = 'https://api.stackexchange.com/2.3/questions'\n",
    "\n",
    "    params = {\n",
    "        'order': 'desc',\n",
    "        'sort': 'votes',\n",
    "        'tagged': tag,\n",
    "        'site': 'stackoverflow',\n",
    "        'fromdate': int(start_date.timestamp()),\n",
    "        'todate': int(end_date.timestamp()),\n",
    "        'min': min_score,\n",
    "        'pagesize': 50  # 30 par défaut, 100 max\n",
    "    }\n",
    "\n",
    "    response = requests.get(api_url, params=params)\n",
    "    data = response.json()\n",
    "\n",
    "    return data\n",
    "\n",
    "# Observons l'activité cette année\n",
    "tag = 'python'\n",
    "start_date = datetime.datetime(2020, 1, 1)\n",
    "end_date = datetime.datetime(2023, 11, 27)\n",
    "\n",
    "result = get_questions(tag, start_date, end_date)\n",
    "display(result)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>creation_date</th>\n",
       "      <th>title</th>\n",
       "      <th>tags</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1595181079</td>\n",
       "      <td>What is pyproject.toml file for?</td>\n",
       "      <td>[python, pip, packaging, pyproject.toml]</td>\n",
       "      <td>343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1607702016</td>\n",
       "      <td>xlrd.biffh.XLRDError: Excel xlsx file; not sup...</td>\n",
       "      <td>[python, pandas, xlrd, pcf]</td>\n",
       "      <td>287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1580572695</td>\n",
       "      <td>What does model.eval() do in pytorch?</td>\n",
       "      <td>[python, machine-learning, deep-learning, pyto...</td>\n",
       "      <td>259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1612516308</td>\n",
       "      <td>ValueError: numpy.ndarray size changed, may in...</td>\n",
       "      <td>[python, pandas, numpy, scikit-learn, python-3.7]</td>\n",
       "      <td>225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1606035195</td>\n",
       "      <td>docker.errors.DockerException: Error while fet...</td>\n",
       "      <td>[python, linux, docker, docker-compose]</td>\n",
       "      <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1593658606</td>\n",
       "      <td>sqlalchemy.exc.NoSuchModuleError: Can&amp;#39;t lo...</td>\n",
       "      <td>[python, postgresql, sqlalchemy, flask-sqlalch...</td>\n",
       "      <td>220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1608242137</td>\n",
       "      <td>Python was not found; run without arguments to...</td>\n",
       "      <td>[python, python-3.x, windows-10]</td>\n",
       "      <td>203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1604676159</td>\n",
       "      <td>DeprecationWarning: executable_path has been d...</td>\n",
       "      <td>[python, selenium]</td>\n",
       "      <td>198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1648058522</td>\n",
       "      <td>How can I fix the &amp;quot;zsh: command not found...</td>\n",
       "      <td>[python, macos, terminal, atom-editor, macos-m...</td>\n",
       "      <td>197</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1587560390</td>\n",
       "      <td>ERROR: Could not build wheels for scipy which ...</td>\n",
       "      <td>[python, scipy, pip, embedded-linux, python-wh...</td>\n",
       "      <td>196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1653965255</td>\n",
       "      <td>TypeError: Descriptors cannot not be created d...</td>\n",
       "      <td>[python, tensorflow, ray]</td>\n",
       "      <td>193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1582109114</td>\n",
       "      <td>How to reinstall python@2 from Homebrew?</td>\n",
       "      <td>[python, macos, homebrew, python-2.x]</td>\n",
       "      <td>191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1606986952</td>\n",
       "      <td>Resolving new pip backtracking runtime issue</td>\n",
       "      <td>[python, pip]</td>\n",
       "      <td>176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1648626375</td>\n",
       "      <td>ImportError: cannot import name &amp;#39;_unicodef...</td>\n",
       "      <td>[python, python-black]</td>\n",
       "      <td>174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1624808189</td>\n",
       "      <td>Understand Python swapping: why is a, b = b, a...</td>\n",
       "      <td>[python, list, indexing, swap]</td>\n",
       "      <td>172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1617722621</td>\n",
       "      <td>Auto-create primary key used when not defining...</td>\n",
       "      <td>[python, python-3.x, django]</td>\n",
       "      <td>171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1607684916</td>\n",
       "      <td>Pandas cannot open an Excel (.xlsx) file</td>\n",
       "      <td>[python, excel, pandas]</td>\n",
       "      <td>171</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1639262655</td>\n",
       "      <td>ImportError: cannot import name &amp;#39;url&amp;#39; ...</td>\n",
       "      <td>[python, django, django-urls, django-4.0]</td>\n",
       "      <td>169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1579523213</td>\n",
       "      <td>Could not load dynamic library &amp;#39;cudart64_1...</td>\n",
       "      <td>[python, python-3.x, tensorflow, keras, tensor...</td>\n",
       "      <td>160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1593001526</td>\n",
       "      <td>How do I install python on alpine linux?</td>\n",
       "      <td>[python, linux, alpine-linux]</td>\n",
       "      <td>155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1677700339</td>\n",
       "      <td>How do I solve &amp;quot;error: externally-managed...</td>\n",
       "      <td>[python, error-handling, pip]</td>\n",
       "      <td>150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1680851159</td>\n",
       "      <td>Error &amp;quot;&amp;#39;DataFrame&amp;#39; object has no ...</td>\n",
       "      <td>[python, pandas, dataframe, attributeerror]</td>\n",
       "      <td>149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1583425357</td>\n",
       "      <td>What causes [*a] to overallocate?</td>\n",
       "      <td>[python, python-3.x, list, cpython, python-int...</td>\n",
       "      <td>149</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1602156410</td>\n",
       "      <td>How to solve &amp;quot;error: Microsoft Visual C++...</td>\n",
       "      <td>[python, python-3.x, visual-studio]</td>\n",
       "      <td>147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1579796188</td>\n",
       "      <td>VSCode doesn&amp;#39;t show poetry virtualenvs in ...</td>\n",
       "      <td>[python, visual-studio-code, python-poetry]</td>\n",
       "      <td>146</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1594683418</td>\n",
       "      <td>pip freeze creates some weird path instead of ...</td>\n",
       "      <td>[python, path, pip, requirements.txt]</td>\n",
       "      <td>144</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>1594067909</td>\n",
       "      <td>How to import requirements.txt from an existin...</td>\n",
       "      <td>[python, pip, python-poetry]</td>\n",
       "      <td>143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1643123383</td>\n",
       "      <td>Does it make sense to use Conda + Poetry?</td>\n",
       "      <td>[python, machine-learning, package, conda, pyt...</td>\n",
       "      <td>142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>1643042804</td>\n",
       "      <td>Good alternative to Pandas .append() method, n...</td>\n",
       "      <td>[python, pandas, dataframe, data-wrangling, da...</td>\n",
       "      <td>142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1604360224</td>\n",
       "      <td>How do you fix &amp;quot;runtimeError: package fai...</td>\n",
       "      <td>[python, python-3.x, windows, numpy]</td>\n",
       "      <td>140</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>1584987372</td>\n",
       "      <td>FastAPI throws an error (Error loading ASGI ap...</td>\n",
       "      <td>[python, fastapi, uvicorn]</td>\n",
       "      <td>137</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>1607990716</td>\n",
       "      <td>What does this tensorflow message mean? Any si...</td>\n",
       "      <td>[python, tensorflow, anaconda]</td>\n",
       "      <td>134</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>1580436978</td>\n",
       "      <td>pip/python: normal site-packages is not writeable</td>\n",
       "      <td>[python, pip]</td>\n",
       "      <td>132</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>1599673099</td>\n",
       "      <td>Youtube_dl : ERROR : YouTube said: Unable to e...</td>\n",
       "      <td>[python, python-3.x, download, youtube, youtub...</td>\n",
       "      <td>131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>1663949925</td>\n",
       "      <td>AttributeError: module &amp;#39;lib&amp;#39; has no at...</td>\n",
       "      <td>[python, google-analytics-api]</td>\n",
       "      <td>128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td>1587382490</td>\n",
       "      <td>Is there a Pathlib alternate for os.path.join?</td>\n",
       "      <td>[python, python-3.x, pathlib]</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>1583716399</td>\n",
       "      <td>ImportError : Attempted relative import with n...</td>\n",
       "      <td>[python, pylint]</td>\n",
       "      <td>123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>1605196086</td>\n",
       "      <td>ImportError: cannot import name &amp;#39;...&amp;#39; ...</td>\n",
       "      <td>[python, django, importerror, python-module, c...</td>\n",
       "      <td>121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>38</th>\n",
       "      <td>1652208276</td>\n",
       "      <td>ImportError: cannot import name &amp;#39;soft_unic...</td>\n",
       "      <td>[python, docker, docker-compose, dependencies,...</td>\n",
       "      <td>117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>1580405248</td>\n",
       "      <td>AttributeError: &amp;#39;DataFrame&amp;#39; object has...</td>\n",
       "      <td>[python, pandas, dataframe]</td>\n",
       "      <td>116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>1633747154</td>\n",
       "      <td>Pip is not working for Python 3.10 on Ubuntu</td>\n",
       "      <td>[python, python-3.x, linux, pip]</td>\n",
       "      <td>115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>1593522500</td>\n",
       "      <td>ConvergenceWarning: lbfgs failed to converge (...</td>\n",
       "      <td>[python, machine-learning, scikit-learn, logis...</td>\n",
       "      <td>115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>42</th>\n",
       "      <td>1611094722</td>\n",
       "      <td>What are &amp;quot;soft keywords&amp;quot;?</td>\n",
       "      <td>[python, python-3.9]</td>\n",
       "      <td>114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>1611639213</td>\n",
       "      <td>Upgrading pip fails with syntax error caused b...</td>\n",
       "      <td>[python, pip]</td>\n",
       "      <td>113</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>1597854368</td>\n",
       "      <td>ModuleNotFoundError: No module named &amp;#39;virt...</td>\n",
       "      <td>[python, virtualenv]</td>\n",
       "      <td>111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>1614021415</td>\n",
       "      <td>Could not install packages due to an OSError: ...</td>\n",
       "      <td>[python, numpy]</td>\n",
       "      <td>110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>1589824360</td>\n",
       "      <td>Ubuntu 20.04 upgrade, Python missing libffi.so.6</td>\n",
       "      <td>[python, ubuntu, ubuntu-20.04]</td>\n",
       "      <td>110</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47</th>\n",
       "      <td>1656337711</td>\n",
       "      <td>Selenium - Python - AttributeError: &amp;#39;WebDr...</td>\n",
       "      <td>[python, selenium, google-chrome, selenium-web...</td>\n",
       "      <td>108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>1651782719</td>\n",
       "      <td>libssl.so.1.1: cannot open shared object file:...</td>\n",
       "      <td>[python, linux, openssl, python-poetry]</td>\n",
       "      <td>107</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>1579148521</td>\n",
       "      <td>How to fix AttributeError: partially initializ...</td>\n",
       "      <td>[python, python-3.x]</td>\n",
       "      <td>107</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    creation_date                                              title  \\\n",
       "0      1595181079                   What is pyproject.toml file for?   \n",
       "1      1607702016  xlrd.biffh.XLRDError: Excel xlsx file; not sup...   \n",
       "2      1580572695              What does model.eval() do in pytorch?   \n",
       "3      1612516308  ValueError: numpy.ndarray size changed, may in...   \n",
       "4      1606035195  docker.errors.DockerException: Error while fet...   \n",
       "5      1593658606  sqlalchemy.exc.NoSuchModuleError: Can&#39;t lo...   \n",
       "6      1608242137  Python was not found; run without arguments to...   \n",
       "7      1604676159  DeprecationWarning: executable_path has been d...   \n",
       "8      1648058522  How can I fix the &quot;zsh: command not found...   \n",
       "9      1587560390  ERROR: Could not build wheels for scipy which ...   \n",
       "10     1653965255  TypeError: Descriptors cannot not be created d...   \n",
       "11     1582109114           How to reinstall python@2 from Homebrew?   \n",
       "12     1606986952       Resolving new pip backtracking runtime issue   \n",
       "13     1648626375  ImportError: cannot import name &#39;_unicodef...   \n",
       "14     1624808189  Understand Python swapping: why is a, b = b, a...   \n",
       "15     1617722621  Auto-create primary key used when not defining...   \n",
       "16     1607684916           Pandas cannot open an Excel (.xlsx) file   \n",
       "17     1639262655  ImportError: cannot import name &#39;url&#39; ...   \n",
       "18     1579523213  Could not load dynamic library &#39;cudart64_1...   \n",
       "19     1593001526           How do I install python on alpine linux?   \n",
       "20     1677700339  How do I solve &quot;error: externally-managed...   \n",
       "21     1680851159  Error &quot;&#39;DataFrame&#39; object has no ...   \n",
       "22     1583425357                  What causes [*a] to overallocate?   \n",
       "23     1602156410  How to solve &quot;error: Microsoft Visual C++...   \n",
       "24     1579796188  VSCode doesn&#39;t show poetry virtualenvs in ...   \n",
       "25     1594683418  pip freeze creates some weird path instead of ...   \n",
       "26     1594067909  How to import requirements.txt from an existin...   \n",
       "27     1643123383          Does it make sense to use Conda + Poetry?   \n",
       "28     1643042804  Good alternative to Pandas .append() method, n...   \n",
       "29     1604360224  How do you fix &quot;runtimeError: package fai...   \n",
       "30     1584987372  FastAPI throws an error (Error loading ASGI ap...   \n",
       "31     1607990716  What does this tensorflow message mean? Any si...   \n",
       "32     1580436978  pip/python: normal site-packages is not writeable   \n",
       "33     1599673099  Youtube_dl : ERROR : YouTube said: Unable to e...   \n",
       "34     1663949925  AttributeError: module &#39;lib&#39; has no at...   \n",
       "35     1587382490     Is there a Pathlib alternate for os.path.join?   \n",
       "36     1583716399  ImportError : Attempted relative import with n...   \n",
       "37     1605196086  ImportError: cannot import name &#39;...&#39; ...   \n",
       "38     1652208276  ImportError: cannot import name &#39;soft_unic...   \n",
       "39     1580405248  AttributeError: &#39;DataFrame&#39; object has...   \n",
       "40     1633747154       Pip is not working for Python 3.10 on Ubuntu   \n",
       "41     1593522500  ConvergenceWarning: lbfgs failed to converge (...   \n",
       "42     1611094722                What are &quot;soft keywords&quot;?   \n",
       "43     1611639213  Upgrading pip fails with syntax error caused b...   \n",
       "44     1597854368  ModuleNotFoundError: No module named &#39;virt...   \n",
       "45     1614021415  Could not install packages due to an OSError: ...   \n",
       "46     1589824360   Ubuntu 20.04 upgrade, Python missing libffi.so.6   \n",
       "47     1656337711  Selenium - Python - AttributeError: &#39;WebDr...   \n",
       "48     1651782719  libssl.so.1.1: cannot open shared object file:...   \n",
       "49     1579148521  How to fix AttributeError: partially initializ...   \n",
       "\n",
       "                                                 tags  score  \n",
       "0            [python, pip, packaging, pyproject.toml]    343  \n",
       "1                         [python, pandas, xlrd, pcf]    287  \n",
       "2   [python, machine-learning, deep-learning, pyto...    259  \n",
       "3   [python, pandas, numpy, scikit-learn, python-3.7]    225  \n",
       "4             [python, linux, docker, docker-compose]    221  \n",
       "5   [python, postgresql, sqlalchemy, flask-sqlalch...    220  \n",
       "6                    [python, python-3.x, windows-10]    203  \n",
       "7                                  [python, selenium]    198  \n",
       "8   [python, macos, terminal, atom-editor, macos-m...    197  \n",
       "9   [python, scipy, pip, embedded-linux, python-wh...    196  \n",
       "10                          [python, tensorflow, ray]    193  \n",
       "11              [python, macos, homebrew, python-2.x]    191  \n",
       "12                                      [python, pip]    176  \n",
       "13                             [python, python-black]    174  \n",
       "14                     [python, list, indexing, swap]    172  \n",
       "15                       [python, python-3.x, django]    171  \n",
       "16                            [python, excel, pandas]    171  \n",
       "17          [python, django, django-urls, django-4.0]    169  \n",
       "18  [python, python-3.x, tensorflow, keras, tensor...    160  \n",
       "19                      [python, linux, alpine-linux]    155  \n",
       "20                      [python, error-handling, pip]    150  \n",
       "21        [python, pandas, dataframe, attributeerror]    149  \n",
       "22  [python, python-3.x, list, cpython, python-int...    149  \n",
       "23                [python, python-3.x, visual-studio]    147  \n",
       "24        [python, visual-studio-code, python-poetry]    146  \n",
       "25              [python, path, pip, requirements.txt]    144  \n",
       "26                       [python, pip, python-poetry]    143  \n",
       "27  [python, machine-learning, package, conda, pyt...    142  \n",
       "28  [python, pandas, dataframe, data-wrangling, da...    142  \n",
       "29               [python, python-3.x, windows, numpy]    140  \n",
       "30                         [python, fastapi, uvicorn]    137  \n",
       "31                     [python, tensorflow, anaconda]    134  \n",
       "32                                      [python, pip]    132  \n",
       "33  [python, python-3.x, download, youtube, youtub...    131  \n",
       "34                     [python, google-analytics-api]    128  \n",
       "35                      [python, python-3.x, pathlib]    123  \n",
       "36                                   [python, pylint]    123  \n",
       "37  [python, django, importerror, python-module, c...    121  \n",
       "38  [python, docker, docker-compose, dependencies,...    117  \n",
       "39                        [python, pandas, dataframe]    116  \n",
       "40                   [python, python-3.x, linux, pip]    115  \n",
       "41  [python, machine-learning, scikit-learn, logis...    115  \n",
       "42                               [python, python-3.9]    114  \n",
       "43                                      [python, pip]    113  \n",
       "44                               [python, virtualenv]    111  \n",
       "45                                    [python, numpy]    110  \n",
       "46                     [python, ubuntu, ubuntu-20.04]    110  \n",
       "47  [python, selenium, google-chrome, selenium-web...    108  \n",
       "48            [python, linux, openssl, python-poetry]    107  \n",
       "49                               [python, python-3.x]    107  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Plus qu'à filtrer / créer le dataframe\n",
    "\n",
    "def extract_fields_and_create_dataframe(data):\n",
    "    questions = data.get('items', [])\n",
    "\n",
    "    # Extracting desired fields\n",
    "    extracted_data = [\n",
    "        {\n",
    "            'creation_date': question.get('creation_date', None),\n",
    "            'title': question.get('title', None),\n",
    "            'tags': question.get('tags', None),\n",
    "            'score': question.get('score', None),\n",
    "        }\n",
    "        for question in questions\n",
    "    ]\n",
    "\n",
    "    # Creating DataFrame\n",
    "    df = pd.DataFrame(extracted_data)\n",
    "\n",
    "    return df\n",
    "\n",
    "df = extract_fields_and_create_dataframe(result)\n",
    "\n",
    "# Displaying the DataFrame\n",
    "display(df)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Préparation de la requete SQL : consultation de la doc (schema)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "# La page d'accueil du \"StackExchange Data Explorer\" (l'outil d’export de données)\n",
    "# présente la structure de la table, avec le type de chaque colonne, et un lien vers la doc :\n",
    "# https://meta.stackexchange.com/questions/2677/database-schema-documentation-for-the-public-data-dump-and-sede\n",
    "\n",
    "# Table : posts\n",
    "# \"You find in Posts all non-deleted posts.\"\n",
    "\n",
    "# La requete donnee en exemple est un tres bon pt de depart.\n",
    "# Tags1 - question tags (PostTypeId = 1)\n",
    "# AnswerCount - the number of undeleted answers (only present if PostTypeId = 1)\n",
    "# FavoriteCount (nullable)\n",
    "\n",
    "# On peut aussi y ajouter :\n",
    "\n",
    "# extraire des questions relativement recentes : par exemple, parce que de nouveaux tags on pu apparaitre,\n",
    "# qui seraient sous-estimes autrement. (c'est un domaine en ébullition !)\n",
    "# CreationDate > ...\n",
    "\n",
    "\n",
    "\n",
    "# autres pistes possibles, mais pas tres utiles je pense :\n",
    "\n",
    "# CommentCount (nullable) : semble redondant puisqu'on a deja AnswerCount et FavoriteCount\n",
    "# de + sur stack il y a generalemt bcp - de commentaires que de reponses\n",
    "# ClosedDate1 (present only if the post is closed) :  pas sur que cela nous importe directement\n",
    "# CommunityOwnedDate (present only if post is community wiki'd) donc pas pour les questions\n",
    "# ContentLicense1 pas pertinent pour nous.\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Requete SQL\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "# SELECT TOP 50000 Title, Body, Tags, Score, ViewCount, FavoriteCount, AnswerCount\n",
    "# FROM Posts\n",
    "# WHERE PostTypeId = 1 AND ViewCount > 10 AND FavoriteCount > 10\n",
    "# AND Score > 5 AND AnswerCount > 0 AND LEN(Tags) - LEN(REPLACE(Tags, '<','')) >= 5\n",
    "# AND CreationDate >= '2020-01-01';\n",
    "\n",
    "# est-ce qu on a vraimt besoin d'extraire les counts ?\n",
    "# pourquoi pas juste Title, Body, Tags ?\n",
    "\n",
    "# 0 row lol\n",
    "\n",
    "# Pas besoin de GROUP BY ou d'ORDER BY ici.\n",
    "\n",
    "# Mais pour s'amuser un peu, on peut choisir d'extraire, parmi nos questions qui remplissent nos criteres,\n",
    "# les plus recentes. Notre requete devient :\n",
    "\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "myenv_p4",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
